<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A guide to plot aesthetics • BentoBox</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="A guide to plot aesthetics">
<meta property="og:description" content="BentoBox">
<meta property="og:image" content="https://phanstiellab.github.io/bentobox/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">BentoBox</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/guides/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../../articles/introduction_to_bentobox.html">Introduction to BentoBox</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/guides/bentobox_page.html">The BentoBox Page</a>
    </li>
    <li>
      <a href="../../articles/guides/plot_elements.html">Additional plot types and elements</a>
    </li>
    <li>
      <a href="../../articles/guides/annotations.html">An in depth look at plot annotations</a>
    </li>
    <li>
      <a href="../../articles/guides/aesthetics.html">A guide to plot aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhanstielLab/BentoBox">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="aesthetics_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A guide to plot aesthetics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PhanstielLab/BentoBox/blob/master/vignettes/guides/aesthetics.Rmd"><code>vignettes/guides/aesthetics.Rmd</code></a></small>
      <div class="hidden name"><code>aesthetics.Rmd</code></div>

    </div>

    
    
<p><code>BentoBox</code> plots are extremely customizable in appearance. In this article we will outline some of the major aesthetic customizations, including general features and specific plot type customizations.</p>
<p>All the data included in this article can be found in the supplementary package <code>BentoBoxData</code>.</p>
<div id="gpar-and-common-plot-customizations" class="section level1">
<h1 class="hasAnchor">
<a href="#gpar-and-common-plot-customizations" class="anchor"></a><code>gpar</code> and common plot customizations</h1>
<p>The most common types of customizations are inherited from <code>grid</code> <code>gpar</code> options. If a function accepts <code>...</code>, this usually refers to <code>gpar</code> options that are not explicity listed as parameters in the function documentation. General valid parameters include:</p>
<table class="table">
<colgroup>
<col width="25%">
<col width="74%">
</colgroup>
<tbody>
<tr class="odd">
<td align="right"><strong>alpha</strong></td>
<td align="left">Alpha channel for transparency (number between 0 and 1).</td>
</tr>
<tr class="even">
<td align="right"><strong>fill</strong></td>
<td align="left">Fill color.</td>
</tr>
<tr class="odd">
<td align="right"><strong>linecolor</strong></td>
<td align="left">Line color.</td>
</tr>
<tr class="even">
<td align="right"><strong>lty</strong></td>
<td align="left">Line type. (0=blank, 1=solid, 2=dashed, 3=dotted, 4=dotdash, 5=longdash, 6=twodash).</td>
</tr>
<tr class="odd">
<td align="right"><strong>lwd</strong></td>
<td align="left">Line width.</td>
</tr>
<tr class="even">
<td align="right"><strong>lineend</strong></td>
<td align="left">Line end style (round, butt, square).</td>
</tr>
<tr class="odd">
<td align="right"><strong>linejoin</strong></td>
<td align="left">Line join style (round, mitre, bevel).</td>
</tr>
<tr class="even">
<td align="right"><strong>linemitre</strong></td>
<td align="left">Line mitre limit (number greater than 1).</td>
</tr>
<tr class="odd">
<td align="right"><strong>fontsize</strong></td>
<td align="left">The size of text, in points.</td>
</tr>
<tr class="even">
<td align="right"><strong>fontcolor</strong></td>
<td align="left">Text color.</td>
</tr>
<tr class="odd">
<td align="right"><strong>fontface</strong></td>
<td align="left">The font face (plain, bold, italic, bold.italic, oblique).</td>
</tr>
<tr class="even">
<td align="right"><strong>fontfamily</strong></td>
<td align="left">The font family.</td>
</tr>
<tr class="odd">
<td align="right"><strong>cex</strong></td>
<td align="left">Scaling multiplier applied to symbols.</td>
</tr>
<tr class="even">
<td align="right"><strong>pch</strong></td>
<td align="left">Plotting character, or symbol (integer codes range from 0 to 25).</td>
</tr>
</tbody>
</table>
<p>Additional fonts for the <code>fontfamily</code> argument can be imported with the packages <code>extrafont</code> and <code>showtext</code>. This makes it possible to incorporate special fonts like Times New Roman, Arial, etc. into <code>BentoBox</code> figures.</p>
</div>
<div id="backgrounds-and-baselines" class="section level1">
<h1 class="hasAnchor">
<a href="#backgrounds-and-baselines" class="anchor"></a>Backgrounds and baselines</h1>
<p>By default, <code>BentoBox</code> plots have transparent backgrounds when placed on a <code>bb_page</code>. In many functions, this background color can be changed with the parameter <code>bg</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr8"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">2000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    bg <span class="op">=</span> <span class="st">"#f6f6f6"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">0.5</span>, width <span class="op">=</span> <span class="fl">2</span>, height <span class="op">=</span> <span class="fl">1</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_01-1.png" width="750" style="display: block; margin: auto;"> This makes it easy to clearly see the precise dimensional boundaries of <code>BentoBox</code> plots.</p>
<p>Some plots also benefit from baselines to quickly show where y = 0. This can aid in data interpretation and guide plot annotation placement. Baselines can be plotted in selective plots with <code>baseline = TRUE</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_plotRanges.html">bb_plotRanges</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_ChIP_CTCF_reads</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">29073000</span>, chromend <span class="op">=</span> <span class="fl">29074000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    fill <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"#7ecdbb"</span>, <span class="st">"#37a7db"</span><span class="op">)</span>,
    baseline <span class="op">=</span> <span class="cn">TRUE</span>, baseline.color <span class="op">=</span> <span class="st">"black"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">6.5</span>, height <span class="op">=</span> <span class="fl">4.25</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_02-1.png" width="1425" style="display: block; margin: auto;"></p>
</div>
<div id="colorby" class="section level1">
<h1 class="hasAnchor">
<a href="#colorby" class="anchor"></a><code>colorby</code>
</h1>
<p>The <code>colorby</code> constructor allows us to color the data elements in <code>BentoBox</code> plots by various data features. These features can be a numerical range, like some kind of score value, or categorical values, like positive or negative strand. The <code>colorby</code> object is constructed by specifying the name of the data column to color by, an optional color palette function, and an optional range for numerical values. If not specified, <code>BentoBox</code> will use the <code>RColorBrewer</code> “YlGnBl” palette for mapping numerical data and the “Pairs” palette for qualitative data.</p>
<p>For example, if we revist the BED plot above, <code>IMR90_ChIP_CTCF_reads</code> has an additional <code>strand</code> column for each BED element:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_CTCF_reads"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">IMR90_ChIP_CTCF_reads</span><span class="op">)</span>
<span class="co">#&gt;          chrom    start      end strand</span>
<span class="co">#&gt; 15554862 chr21 28000052 28000088      -</span>
<span class="co">#&gt; 15554863 chr21 28000092 28000128      -</span>
<span class="co">#&gt; 15554864 chr21 28000162 28000198      -</span>
<span class="co">#&gt; 15554865 chr21 28000251 28000287      +</span>
<span class="co">#&gt; 15554866 chr21 28000335 28000371      -</span>
<span class="co">#&gt; 15554867 chr21 28000500 28000536      +</span></code></pre></div>
<p>Thus, we can use the <code>colorby</code> constructor to color BED elements by positive or negative strand. The <code>strand</code> column will be converted to a factor with a <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> level and <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> level. These values will be mapped to our input <code>palette</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">IMR90_ChIP_CTCF_reads</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../../reference/bb_plotRanges.html">bb_plotRanges</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_ChIP_CTCF_reads</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">29073000</span>, chromend <span class="op">=</span> <span class="fl">29074000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    fill <span class="op">=</span> <span class="fu"><a href="../../reference/colorby.html">colorby</a></span><span class="op">(</span><span class="st">"strand"</span>, palette <span class="op">=</span> 
                       <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"#7ecdbb"</span>, <span class="st">"#37a7db"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">6.5</span>, height <span class="op">=</span> <span class="fl">4.25</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_03-1.png" width="1425" style="display: block; margin: auto;"></p>
<p>To further control the order of color mapping, we can set our categorical <code>colorby</code> column as a factor with our own order of levels before plotting:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_CTCF_reads"</span><span class="op">)</span>
<span class="va">IMR90_ChIP_CTCF_reads</span><span class="op">$</span><span class="va">strand</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">IMR90_ChIP_CTCF_reads</span><span class="op">$</span><span class="va">strand</span>, levels <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">IMR90_ChIP_CTCF_reads</span><span class="op">$</span><span class="va">strand</span><span class="op">)</span>
<span class="co">#&gt; [1] - - - + - +</span>
<span class="co">#&gt; Levels: + -</span></code></pre></div>
<p>Now we’ve set the <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> level as our first level, so our <code>palette</code> will map colors in the opposite order from before:</p>
<p><img src="aesthetics_files/figure-html/aes_04-1.png" width="1425" style="display: block; margin: auto;"></p>
<p>In this example, we will color BEDPE arches by a range of numerical values we will add as a <code>length</code> column:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_DNAloops_pairs"</span><span class="op">)</span>
<span class="va">IMR90_DNAloops_pairs</span><span class="op">$</span><span class="va">length</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">IMR90_DNAloops_pairs</span><span class="op">$</span><span class="va">start2</span> <span class="op">-</span> <span class="va">IMR90_DNAloops_pairs</span><span class="op">$</span><span class="va">start1</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1000</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">IMR90_DNAloops_pairs</span><span class="op">$</span><span class="va">length</span><span class="op">)</span>
<span class="co">#&gt; [1]   65 1960 2100  850 1200 1485</span></code></pre></div>
<p>Now we can set <code>fill</code> as a <code>colorby</code> object to color the BEDPE <code>length</code> column by:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bedpePlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotPairsArches.html">bb_plotPairsArches</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_DNAloops_pairs</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">27900000</span>, chromend <span class="op">=</span> <span class="fl">30700000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    fill <span class="op">=</span> <span class="fu"><a href="../../reference/colorby.html">colorby</a></span><span class="op">(</span><span class="st">"length"</span>, 
                   palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"dodgerblue2"</span>, <span class="st">"firebrick2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
    linecolor <span class="op">=</span> <span class="st">"fill"</span>,
    archHeight <span class="op">=</span> <span class="va">IMR90_DNAloops_pairs</span><span class="op">$</span><span class="va">length</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">IMR90_DNAloops_pairs</span><span class="op">$</span><span class="va">length</span><span class="op">)</span>,
    alpha <span class="op">=</span> <span class="fl">1</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">7</span>, height <span class="op">=</span> <span class="fl">1.5</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_05-1.png" width="1425" style="display: block; margin: auto;"></p>
<p>And now since we have numbers mapped to colors, we can use <code><a href="../../reference/bb_annoHeatmapLegend.html">bb_annoHeatmapLegend()</a></code> with our <code>bb_arches</code> object to add a legend for the <code>colorby</code> we performed:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_annoHeatmapLegend.html">bb_annoHeatmapLegend</a></span><span class="op">(</span>
    plot <span class="op">=</span> <span class="va">bedpePlot</span>, fontcolor <span class="op">=</span> <span class="st">"black"</span>,
    x <span class="op">=</span> <span class="fl">7.0</span>, y <span class="op">=</span> <span class="fl">0.25</span>,
    width <span class="op">=</span> <span class="fl">0.10</span>, height <span class="op">=</span> <span class="fl">1</span>, fontsize <span class="op">=</span> <span class="fl">10</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_06-1.png" width="1425" style="display: block; margin: auto;"></p>
</div>
<div id="gene-and-transcript-plot-aesthetics" class="section level1">
<h1 class="hasAnchor">
<a href="#gene-and-transcript-plot-aesthetics" class="anchor"></a>Gene and transcript plot aesthetics</h1>
<p><code>BentoBox</code> provides many useful features specific for enhancing gene and transcript visualizations:</p>
<div id="labels" class="section level2">
<h2 class="hasAnchor">
<a href="#labels" class="anchor"></a>Labels</h2>
<p>Since <code>BentoBox</code> utilizes <code>TxDb</code> objects, <code>orgDb</code> objects, and internal citation information, <code>BentoBox</code> has access to numerous gene and transcript identifiers and can customize annotation labels in a variety of ways.</p>
<p>By default, <code>BentoBox</code> will rank gene labels according to citation number to prevent label overcrowding. However, we can provide our own list of prioritized genes to label in a plot. For example, if we plot the hg19 genes in the region <code>chr2:1000000-20000000</code>, our plot will show these labels:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">1.25</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">genePlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">4.75</span>, height <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_07-1.png" width="750" style="display: block; margin: auto;"> Looking in the <code>bb_genes</code> object, we can see that there were numerous genes that were not labeled.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genePlot</span><span class="op">$</span><span class="va">genes</span>
<span class="co">#&gt;  [1] "MIR3125"      "RNASEH1-AS1"  "LOC100506274" "MIR4757"      "LINC00570"   </span>
<span class="co">#&gt;  [6] "C2orf50"      "SLC66A3"      "SILC1"        "LRATD1"       "DDX1"        </span>
<span class="co">#&gt; [11] "LPIN1"        "ATP6V1C2"     "IAH1"         "TRIB2"        "GRHL1"       </span>
<span class="co">#&gt; [16] "HPCAL1"       "ID2"          "MSGN1"        "GEN1"         NA            </span>
<span class="co">#&gt; [21] "KCNF1"        "KCNS3"        "LOC400940"    "MYCN"         "TRAPPC12"    </span>
<span class="co">#&gt; [26] "CPSF3"        "SNTG2"        "ALLC"         "RPS7"         "RRM2"        </span>
<span class="co">#&gt; [31] "SOX11"        "TPO"          "MYT1L-AS1"    "VSNL1"        "COLEC11"     </span>
<span class="co">#&gt; [36] "KLF11"        "ASAP2"        "TAF1B"        "RSAD2"        "GREB1"       </span>
<span class="co">#&gt; [41] "RNF144A"      "SNORA80B"     "MIR4261"      "MIR4262"      "LINC01304"   </span>
<span class="co">#&gt; [46] "LOC100506474" "NT5C1B-RDH14" "MIR4429"      "PDIA6"        "MYCNOS"      </span>
<span class="co">#&gt; [51] "YWHAQ"        "CMPK2"        "MBOAT2"       "OSR1"         "E2F6"        </span>
<span class="co">#&gt; [56] "CYS1"         "MYT1L"        "NTSR2"        "RNASEH1"      "FLJ33534"    </span>
<span class="co">#&gt; [61] "LINC00298"    "LINC00299"    "GRASLND"      "LINC00487"    "ODC1"        </span>
<span class="co">#&gt; [66] "NBAS"         "ADI1"         "KIDINS220"    "RDH14"        "ADAM17"      </span>
<span class="co">#&gt; [71] "EIPR1"        "LINC01249"    "RAD51AP2"     "PXDN"         "SMC6"        </span>
<span class="co">#&gt; [76] "NOL10"        "CYRIA"        "ITGB1BP1"     "NT5C1B"       "ROCK2"</span></code></pre></div>
<p>If we were particularly interested in MIR3125, we could include this in the <code>geneOrder</code> parameter to prioritize its labeling:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">1.25</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">genePlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    geneOrder <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"MIR3125"</span><span class="op">)</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">4.75</span>, height <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_08-1.png" width="750" style="display: block; margin: auto;"></p>
<p>If we wanted to change the type of label in our plot, we can change the <code>display.column</code> parameter in a <code><a href="../../reference/bb_assembly.html">bb_assembly()</a></code> object. The default <code>display.column = "SYMBOL"</code>, but we could change this value to other available <code>keytypes</code> in the <code>orgDb</code> we are using. For example, if we wanted to display the associated Ensembl IDs, we would set <code>display.column = "ENSEMBL"</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">new_hg19</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_assembly.html">bb_assembly</a></span><span class="op">(</span>
    Genome <span class="op">=</span> <span class="st">"id_hg19"</span>,
    TxDb <span class="op">=</span> <span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span>,
    OrgDb <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>,
    gene.id.column <span class="op">=</span> <span class="st">"ENTREZID"</span>,
    display.column <span class="op">=</span> <span class="st">"ENSEMBL"</span>
<span class="op">)</span>
<span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">1.25</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">genePlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,
    assembly <span class="op">=</span> <span class="va">new_hg19</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">4.75</span>, height <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_09-1.png" width="750" style="display: block; margin: auto;"></p>
<p>Label IDs used in transcript plots can be customized through <code><a href="../../reference/bb_assembly.html">bb_assembly()</a></code> objects, and transcript label formatting can be changed through the <code>labels</code> parameter. For example, if we wish to display both transcript names and their associated gene names, we can set <code>labels = "both"</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">6</span>, height <span class="op">=</span> <span class="fl">5</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">transcriptPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotTranscripts.html">bb_plotTranscripts</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    labels <span class="op">=</span> <span class="st">"both"</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">5.5</span>, height <span class="op">=</span> <span class="fl">4.5</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_10-1.png" width="1050" style="display: block; margin: auto;"></p>
</div>
<div id="highlighting-genes-by-color" class="section level2">
<h2 class="hasAnchor">
<a href="#highlighting-genes-by-color" class="anchor"></a>Highlighting genes by color</h2>
<p>In addition to changing <code>fill</code> and <code>fontcolor</code> to change the colors of all genes in a plot, specific gene structures and their labels can be highlighted in a different color with <code>geneHighlights</code>. If we revisit the <code>bb_genes</code> plot above, we can highlight RRM2 by creating a <code>data.frame</code> with “RRM2” in the first column and its highlight color in the second column:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneHighlights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"geneName"</span> <span class="op">=</span> <span class="st">"RRM2"</span>, <span class="st">"color"</span> <span class="op">=</span> <span class="st">"steel blue"</span><span class="op">)</span></code></pre></div>
<p>We can then pass this into our <code><a href="../../reference/bb_plotGenes.html">bb_plotGenes()</a></code> call:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">1.25</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">genePlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    geneHighlights <span class="op">=</span> <span class="va">geneHighlights</span>, geneBackground <span class="op">=</span> <span class="st">"grey"</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">4.75</span>, height <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_11-1.png" width="750" style="display: block; margin: auto;"></p>
<p>Since <code>geneHighlights</code> is a <code>data.frame</code>, we can highlight multiple genes in different colors at once. For example, let’s now highlight RRM2 in “steel blue” and PXDN in “red”:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneHighlights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
    <span class="st">"geneName"</span> <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"RRM2"</span>, <span class="st">"PXDN"</span><span class="op">)</span>,
    <span class="st">"color"</span> <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"steel blue"</span>, <span class="st">"red"</span><span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">1.25</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">genePlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotGenes.html">bb_plotGenes</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    geneHighlights <span class="op">=</span> <span class="va">geneHighlights</span>, geneBackground <span class="op">=</span> <span class="st">"grey"</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">4.75</span>, height <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_12-1.png" width="750" style="display: block; margin: auto;"></p>
</div>
<div id="customizing-transcripts-by-strand" class="section level2">
<h2 class="hasAnchor">
<a href="#customizing-transcripts-by-strand" class="anchor"></a>Customizing transcripts by strand</h2>
<p>To distinguish which strand a transcript belongs to, <code><a href="../../reference/bb_plotTranscripts.html">bb_plotTranscripts()</a></code> colors transcripts by strand with the parameter <code>colorbyStrand</code>. The first value in <code>fill</code> colors positive strand transcripts and the second <code>fill</code> value colors negative strand transcripts. To further organize transcripts by strand, we can use <code>strandSplit</code> to separate transcript elements into groups of positive and negative strands:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">6</span>, height <span class="op">=</span> <span class="fl">5</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">transcriptPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotTranscripts.html">bb_plotTranscripts</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    strandSplit <span class="op">=</span> <span class="cn">TRUE</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">5.5</span>, height <span class="op">=</span> <span class="fl">4.5</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_13-1.png" width="1050" style="display: block; margin: auto;"></p>
<p>Now all our positive strand transcripts are grouped together above the group of negative strand transcripts.</p>
</div>
</div>
<div id="hi-c-plot-customizations" class="section level1">
<h1 class="hasAnchor">
<a href="#hi-c-plot-customizations" class="anchor"></a>Hi-C plot customizations</h1>
<p><code>BentoBox</code> includes many types of customizations for Hi-C plots. <code>BentoBox</code> provides 3 different Hi-C plotting functions based on the desired plot shape:</p>
<ul>
<li><p><code><a href="../../reference/bb_plotHicSquare.html">bb_plotHicSquare()</a></code>: Plots a square, symmetrical Hi-C plot with genomic coordinates along both the x- and y-axes.</p></li>
<li><p><code><a href="../../reference/bb_plotHicTriangle.html">bb_plotHicTriangle()</a></code>: Plots a triangular Hi-C plot where the genome region falls along the base of the triangle.</p></li>
<li><p><code><a href="../../reference/bb_plotHicRectangle.html">bb_plotHicRectangle()</a></code>: Plots a triangular Hi-C plot with additional data filling the surrounding regions to form a rectangle.</p></li>
</ul>
<p><img src="aesthetics_files/figure-html/aes_14-1.png" width="1462.5" style="display: block; margin: auto;"></p>
<p>All Hi-C plot types can use different color palettes, and colors can be linearly or log-scaled with the <code>colorTrans</code> parameter.</p>
<p><code>bb_hicSquare</code> plots can be further customized to include two datasets in one plot. Instead of plotting both symmetrical halves of the plot, we can set one dataset as <code>half = "top"</code> and the other dataset as <code>half = "bottom"</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GM12878_HiC_10kb"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span>

<span class="fu"><a href="../../reference/bb_pageCreate.html">bb_pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>
<span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_params.html">bb_params</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>, resolution <span class="op">=</span> <span class="fl">10000</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">2.75</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>


<span class="va">hicPlot_top</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotHicSquare.html">bb_plotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">GM12878_HiC_10kb</span>, params <span class="op">=</span> <span class="va">params</span>,
    zrange <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">200</span><span class="op">)</span>,
    half <span class="op">=</span> <span class="st">"top"</span>,
    y <span class="op">=</span> <span class="fl">0.25</span>, height <span class="op">=</span> <span class="fl">2.75</span>
<span class="op">)</span>
<span class="va">hicPlot_bottom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/bb_plotHicSquare.html">bb_plotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>, params <span class="op">=</span> <span class="va">params</span>,
    zrange <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">70</span><span class="op">)</span>,
    half <span class="op">=</span> <span class="st">"bottom"</span>,
    y <span class="op">=</span> <span class="fl">0.25</span>, height <span class="op">=</span> <span class="fl">2.75</span>
<span class="op">)</span></code></pre></div>
<p><img src="aesthetics_files/figure-html/aes_15-1.png" width="787.5" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/nekramer">Nicole Kramer</a>, <a href="https://github.com/EricSDavis">Eric S. Davis</a>, <a href="https://github.com/cwenger">Craig Wenger</a>, <a href="https://github.com/dphansti">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
