<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bioconductor Integration • plotgardener</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Bioconductor Integration">
<meta property="og:description" content="plotgardener">
<meta property="og:image" content="https://phanstiellab.github.io/plotgardener/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">plotgardener</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/guides/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../../articles/introduction_to_plotgardener.html">Introduction to plotgardener</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/guides/plotgardener_page.html">The plotgardener Page</a>
    </li>
    <li>
      <a href="../../articles/guides/reading_data_for_plotgardener.html">Reading Data for plotgardener</a>
    </li>
    <li>
      <a href="../../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../../articles/guides/plotgardener_meta_functions.html">plotgardener Meta Functions</a>
    </li>
    <li>
      <a href="../../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../../articles/guides/incorporating_ggplots.html">Incorporating ggplots and other grid-based Bioconductor visualizations</a>
    </li>
    <li>
      <a href="../../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhanstielLab/plotgardener" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Bioconductor Integration</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PhanstielLab/plotgardener/blob/HEAD/vignettes/guides/bioconductor_integration.Rmd" class="external-link"><code>vignettes/guides/bioconductor_integration.Rmd</code></a></small>
      <div class="hidden name"><code>bioconductor_integration.Rmd</code></div>

    </div>

    
    
<p><code>plotgardener</code> is designed to be flexibly compatible with
typical Bioconductor classes and libraries of genomic data to easily
integrate genomic data analysis and visualization. In addition to
handling various genomic file types and <code>R</code> objects, many
<code>plotgardener</code> functions can also handle <code>GRanges</code>
or <code>GInteractions</code> objects as input data. Furthermore,
<code>plotgardener</code> does not hard-code any genomic assemblies and
can utilize <code>TxDb</code>, <code>OrgDb</code>, and
<code>BSgenome</code> objects for various genomic annotations, including
gene and transcript structures and names, chromosome sizes, and
nucleotide sequences. Furthermore, all cytoband data for ideogram plots
is retrieved from the UCSC Genome Browser through
<code>AnnotationHub</code>. For standard genomic assemblies (i.e. hg19,
hg38, mm10), <code>plotgardener</code> uses a set of default packages
that can be displayed by calling <code><a href="../../reference/defaultPackages.html">defaultPackages()</a></code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/defaultPackages.html">defaultPackages</a></span><span class="op">(</span><span class="st">"hg38"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    1 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;  $ Genome        : chr "hg38"</span></span>
<span><span class="co">#&gt;  $ TxDb          : chr "TxDb.Hsapiens.UCSC.hg38.knownGene"</span></span>
<span><span class="co">#&gt;  $ OrgDb         : chr "org.Hs.eg.db"</span></span>
<span><span class="co">#&gt;  $ gene.id.column: chr "ENTREZID"</span></span>
<span><span class="co">#&gt;  $ display.column: chr "SYMBOL"</span></span>
<span><span class="co">#&gt;  $ BSgenome      : chr "BSgenome.Hsapiens.UCSC.hg38"</span></span>
<span><span class="fu"><a href="../../reference/defaultPackages.html">defaultPackages</a></span><span class="op">(</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    1 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;  $ Genome        : chr "hg19"</span></span>
<span><span class="co">#&gt;  $ TxDb          : chr "TxDb.Hsapiens.UCSC.hg19.knownGene"</span></span>
<span><span class="co">#&gt;  $ OrgDb         : chr "org.Hs.eg.db"</span></span>
<span><span class="co">#&gt;  $ gene.id.column: chr "ENTREZID"</span></span>
<span><span class="co">#&gt;  $ display.column: chr "SYMBOL"</span></span>
<span><span class="co">#&gt;  $ BSgenome      : chr "BSgenome.Hsapiens.UCSC.hg19"</span></span>
<span><span class="fu"><a href="../../reference/defaultPackages.html">defaultPackages</a></span><span class="op">(</span><span class="st">"mm10"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    1 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;  $ Genome        : chr "mm10"</span></span>
<span><span class="co">#&gt;  $ TxDb          : chr "TxDb.Mmusculus.UCSC.mm10.knownGene"</span></span>
<span><span class="co">#&gt;  $ OrgDb         : chr "org.Mm.eg.db"</span></span>
<span><span class="co">#&gt;  $ gene.id.column: chr "ENTREZID"</span></span>
<span><span class="co">#&gt;  $ display.column: chr "SYMBOL"</span></span>
<span><span class="co">#&gt;  $ BSgenome      : chr "BSgenome.Mmusculus.UCSC.mm10"</span></span></code></pre></div>
<p>To see which assemblies have defaults within
<code>plotgardener</code>, call <code><a href="../../reference/genomes.html">genomes()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/genomes.html">genomes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; bosTau8</span></span>
<span><span class="co">#&gt; bosTau9</span></span>
<span><span class="co">#&gt; canFam3</span></span>
<span><span class="co">#&gt; ce6</span></span>
<span><span class="co">#&gt; ce11</span></span>
<span><span class="co">#&gt; danRer10</span></span>
<span><span class="co">#&gt; danRer11</span></span>
<span><span class="co">#&gt; dm3</span></span>
<span><span class="co">#&gt; dm6</span></span>
<span><span class="co">#&gt; galGal4</span></span>
<span><span class="co">#&gt; galGal5</span></span>
<span><span class="co">#&gt; galGal6</span></span>
<span><span class="co">#&gt; hg18</span></span>
<span><span class="co">#&gt; hg19</span></span>
<span><span class="co">#&gt; hg38</span></span>
<span><span class="co">#&gt; mm9</span></span>
<span><span class="co">#&gt; mm10</span></span>
<span><span class="co">#&gt; rheMac3</span></span>
<span><span class="co">#&gt; rheMac8</span></span>
<span><span class="co">#&gt; rheMac10</span></span>
<span><span class="co">#&gt; panTro5</span></span>
<span><span class="co">#&gt; panTro6</span></span>
<span><span class="co">#&gt; rn4</span></span>
<span><span class="co">#&gt; rn5</span></span>
<span><span class="co">#&gt; rn6</span></span>
<span><span class="co">#&gt; sacCer2</span></span>
<span><span class="co">#&gt; sacCer3</span></span>
<span><span class="co">#&gt; susScr3</span></span>
<span><span class="co">#&gt; susScr11</span></span></code></pre></div>
<p><code>plotgardener</code> functions default to an “hg38” assembly,
but can be customized with any of the other genomic assemblies included
or a <code>assembly</code> object. To create custom genomic assemblies
and combinations of <code>TxDb</code>, <code>orgDb</code>, and
<code>BSgenome</code> packages for use in <code>plotgardener</code>
functions, we can use the <code><a href="../../reference/assembly.html">assembly()</a></code> constructor. For
example, we can create our own <code>TxDb</code> from the current human
Ensembl release:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span></span>
<span><span class="va">TxDb.Hsapiens.Ensembl.GrCh38.103</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromEnsembl.html" class="external-link">makeTxDbFromEnsembl</a></span><span class="op">(</span></span>
<span>    organism <span class="op">=</span></span>
<span>        <span class="st">"Homo sapiens"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can now create a new <code>assembly</code> with this
<code>TxDb</code> and combinations of other Bioconductor packages. The
<code>Genome</code> parameter can is a string to name or describe this
assembly. Since the <code>TxDb</code> is now from ENSEMBL, we will
change the <code>gene.id</code> field to <code>"ENSEMBL"</code> to map
gene IDs and symbols between our <code>TxDb</code> and
<code>orgDb</code> objects. Most gene ID types can be found by calling
<code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">AnnotationDbi::keytypes()</a></code> on an <code>orgDb</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ensembl38</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/assembly.html">assembly</a></span><span class="op">(</span></span>
<span>    Genome <span class="op">=</span> <span class="st">"Ensembl.GRCh38.103"</span>,</span>
<span>    TxDb <span class="op">=</span> <span class="va">TxDb.Hsapiens.Ensembl.GrCh38.103</span>,</span>
<span>    OrgDb <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>,</span>
<span>    BSgenome <span class="op">=</span> <span class="st">"BSgenome.Hsapiens.NCBI.GRCh38"</span>,</span>
<span>    gene.id <span class="op">=</span> <span class="st">"ENSEMBL"</span>, display.column <span class="op">=</span> <span class="st">"SYMBOL"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This <code>assembly</code> object can now be easily passed into
<code>plotgardener</code> functions through the <code>assembly</code>
parameter to use custom genomic assembly configurations.</p>
<p><code>assembly</code> objects are especially handy for changing the
type of gene or transcript label of our gene and transcript plots. The
default <code>display.column = "SYMBOL"</code>, but we could change this
value to other available <code>keytypes</code> in the <code>orgDb</code>
we are using. For example, if we wanted to display the associated
Ensembl IDs of an “hg19” <code>assembly</code> object, we would set
<code>display.column = "ENSEMBL"</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">new_hg19</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/assembly.html">assembly</a></span><span class="op">(</span></span>
<span>    Genome <span class="op">=</span> <span class="st">"id_hg19"</span>,</span>
<span>    TxDb <span class="op">=</span> <span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span>,</span>
<span>    OrgDb <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>,</span>
<span>    gene.id.column <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>    display.column <span class="op">=</span> <span class="st">"ENSEMBL"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/pageCreate.html">pageCreate</a></span><span class="op">(</span></span>
<span>    width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">1.25</span>,</span>
<span>    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span><span class="va">genePlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotGenes.html">plotGenes</a></span><span class="op">(</span></span>
<span>    chrom <span class="op">=</span> <span class="st">"chr2"</span>, chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">20000000</span>,</span>
<span>    assembly <span class="op">=</span> <span class="va">new_hg19</span>,</span>
<span>    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">4.75</span>, height <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="bioconductor_integration_files/figure-html/bioc_01-1.png" width="750" style="display: block; margin: auto;"></p>
<div class="section level3">
<h3 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    grid      stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] org.Hs.eg.db_3.15.0                    </span></span>
<span><span class="co">#&gt;  [2] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2</span></span>
<span><span class="co">#&gt;  [3] GenomicFeatures_1.48.3                 </span></span>
<span><span class="co">#&gt;  [4] AnnotationDbi_1.58.0                   </span></span>
<span><span class="co">#&gt;  [5] Biobase_2.56.0                         </span></span>
<span><span class="co">#&gt;  [6] GenomicRanges_1.48.0                   </span></span>
<span><span class="co">#&gt;  [7] GenomeInfoDb_1.32.2                    </span></span>
<span><span class="co">#&gt;  [8] IRanges_2.30.0                         </span></span>
<span><span class="co">#&gt;  [9] S4Vectors_0.34.0                       </span></span>
<span><span class="co">#&gt; [10] BiocGenerics_0.42.0                    </span></span>
<span><span class="co">#&gt; [11] plotgardener_1.2.9                     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] bitops_1.0-7                matrixStats_0.62.0         </span></span>
<span><span class="co">#&gt;  [3] fs_1.5.2                    bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [5] filelock_1.0.2              progress_1.2.2             </span></span>
<span><span class="co">#&gt;  [7] RColorBrewer_1.1-3          httr_1.4.3                 </span></span>
<span><span class="co">#&gt;  [9] rprojroot_2.0.3             tools_4.2.0                </span></span>
<span><span class="co">#&gt; [11] bslib_0.4.0                 utf8_1.2.2                 </span></span>
<span><span class="co">#&gt; [13] R6_2.5.1                    DBI_1.1.3                  </span></span>
<span><span class="co">#&gt; [15] colorspace_2.0-3            prettyunits_1.1.1          </span></span>
<span><span class="co">#&gt; [17] tidyselect_1.1.2            bit_4.0.4                  </span></span>
<span><span class="co">#&gt; [19] curl_4.3.2                  compiler_4.2.0             </span></span>
<span><span class="co">#&gt; [21] textshaping_0.3.6           cli_3.3.0                  </span></span>
<span><span class="co">#&gt; [23] xml2_1.3.3                  desc_1.4.1                 </span></span>
<span><span class="co">#&gt; [25] DelayedArray_0.22.0         rtracklayer_1.56.1         </span></span>
<span><span class="co">#&gt; [27] sass_0.4.2                  scales_1.2.0               </span></span>
<span><span class="co">#&gt; [29] rappdirs_0.3.3              pkgdown_2.0.6              </span></span>
<span><span class="co">#&gt; [31] systemfonts_1.0.4           stringr_1.4.0              </span></span>
<span><span class="co">#&gt; [33] digest_0.6.29               Rsamtools_2.12.0           </span></span>
<span><span class="co">#&gt; [35] yulab.utils_0.0.5           rmarkdown_2.14             </span></span>
<span><span class="co">#&gt; [37] XVector_0.36.0              pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [39] htmltools_0.5.3             MatrixGenerics_1.8.1       </span></span>
<span><span class="co">#&gt; [41] highr_0.9                   dbplyr_2.2.1               </span></span>
<span><span class="co">#&gt; [43] fastmap_1.1.0               rlang_1.0.4                </span></span>
<span><span class="co">#&gt; [45] RSQLite_2.2.15              gridGraphics_0.5-1         </span></span>
<span><span class="co">#&gt; [47] jquerylib_0.1.4             BiocIO_1.6.0               </span></span>
<span><span class="co">#&gt; [49] generics_0.1.3              jsonlite_1.8.0             </span></span>
<span><span class="co">#&gt; [51] BiocParallel_1.30.3         dplyr_1.0.9                </span></span>
<span><span class="co">#&gt; [53] RCurl_1.98-1.7              magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [55] ggplotify_0.1.0             GenomeInfoDbData_1.2.8     </span></span>
<span><span class="co">#&gt; [57] Matrix_1.4-1                Rcpp_1.0.9                 </span></span>
<span><span class="co">#&gt; [59] munsell_0.5.0               fansi_1.0.3                </span></span>
<span><span class="co">#&gt; [61] lifecycle_1.0.1             stringi_1.7.8              </span></span>
<span><span class="co">#&gt; [63] yaml_2.3.5                  SummarizedExperiment_1.26.1</span></span>
<span><span class="co">#&gt; [65] zlibbioc_1.42.0             BiocFileCache_2.4.0        </span></span>
<span><span class="co">#&gt; [67] blob_1.2.3                  parallel_4.2.0             </span></span>
<span><span class="co">#&gt; [69] crayon_1.5.1                lattice_0.20-45            </span></span>
<span><span class="co">#&gt; [71] Biostrings_2.64.0           hms_1.1.1                  </span></span>
<span><span class="co">#&gt; [73] KEGGREST_1.36.3             knitr_1.39                 </span></span>
<span><span class="co">#&gt; [75] pillar_1.8.0                rjson_0.2.21               </span></span>
<span><span class="co">#&gt; [77] biomaRt_2.52.0              codetools_0.2-18           </span></span>
<span><span class="co">#&gt; [79] strawr_0.0.9                XML_3.99-0.10              </span></span>
<span><span class="co">#&gt; [81] glue_1.6.2                  evaluate_0.15              </span></span>
<span><span class="co">#&gt; [83] data.table_1.14.2           vctrs_0.4.1                </span></span>
<span><span class="co">#&gt; [85] png_0.1-7                   gtable_0.3.0               </span></span>
<span><span class="co">#&gt; [87] purrr_0.3.4                 assertthat_0.2.1           </span></span>
<span><span class="co">#&gt; [89] cachem_1.0.6                ggplot2_3.3.6              </span></span>
<span><span class="co">#&gt; [91] xfun_0.31                   restfulr_0.0.15            </span></span>
<span><span class="co">#&gt; [93] ragg_1.2.2                  tibble_3.1.8               </span></span>
<span><span class="co">#&gt; [95] GenomicAlignments_1.32.1    plyranges_1.16.0           </span></span>
<span><span class="co">#&gt; [97] memoise_2.0.1               ellipsis_0.3.2</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nekramer" class="external-link">Nicole Kramer</a>, <a href="https://github.com/EricSDavis" class="external-link">Eric S. Davis</a>, <a href="https://github.com/cwenger" class="external-link">Craig Wenger</a>, <a href="https://github.com/dphansti" class="external-link">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
