<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Plotting Multi-omic Data • plotgardener</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><link href="../../extra.css" rel="stylesheet">
<meta property="og:title" content="Plotting Multi-omic Data">
<meta property="og:description" content="plotgardener">
<meta property="og:image" content="https://phanstiellab.github.io/plotgardener/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">plotgardener</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../articles/guides/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../../articles/introduction_to_plotgardener.html">Introduction to plotgardener</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/guides/plotgardener_page.html">The plotgardener Page</a>
    </li>
    <li>
      <a href="../../articles/guides/reading_data_for_plotgardener.html">Reading Data for plotgardener</a>
    </li>
    <li>
      <a href="../../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../../articles/guides/plotgardener_meta_functions.html">plotgardener Meta Functions</a>
    </li>
    <li>
      <a href="../../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../../articles/guides/incorporating_ggplots.html">Incorporating ggplots and other grid-based Bioconductor visualizations</a>
    </li>
    <li>
      <a href="../../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhanstielLab/plotgardener" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Plotting Multi-omic Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PhanstielLab/plotgardener/blob/HEAD/vignettes/guides/plotting_multiomic_data.Rmd" class="external-link"><code>vignettes/guides/plotting_multiomic_data.Rmd</code></a></small>
      <div class="hidden name"><code>plotting_multiomic_data.Rmd</code></div>

    </div>

    
    
<p><code>plotgardener</code> makes it easy to create reproducible, publication-quality figures from multi-omic data. Since each plot can be placed in <em>exactly</em> the desired location, users can stack multiple types of genomic data so that their axes and data are correctly aligned. In this section we will show some examples of plotting multi-omic data and how the <code>pgParams</code> object and “below” y-coordinate can facilitate this process.</p>
<p>In the following example, we plot the same genomic region (i.e. <code>chr21:28000000-30300000</code>) represented in Hi-C data, loop annotations, signal track data, GWAS data, all along a common gene track and genome label axis:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load example data</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/plotgardenerData" class="external-link">plotgardenerData</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_DNAloops_pairs"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_H3K27ac_signal"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"hg19_insulin_GWAS"</span><span class="op">)</span>

<span class="co">## Create a plotgardener page</span>
<span class="fu"><a href="../../reference/pageCreate.html">pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">3</span>, height <span class="op">=</span> <span class="fl">5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>

<span class="co">## Plot Hi-C data in region</span>
<span class="fu"><a href="../../reference/plotHicSquare.html">plotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">0.5</span>, width <span class="op">=</span> <span class="fl">2</span>, height <span class="op">=</span> <span class="fl">2</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="co">## Plot loop annotations</span>
<span class="fu"><a href="../../reference/plotPairsArches.html">plotPairsArches</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_DNAloops_pairs</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">2.5</span>, width <span class="op">=</span> <span class="fl">2</span>, height <span class="op">=</span> <span class="fl">0.25</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>,
    fill <span class="op">=</span> <span class="st">"black"</span>, linecolor <span class="op">=</span> <span class="st">"black"</span>, flip <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## Plot signal track data</span>
<span class="fu"><a href="../../reference/plotSignal.html">plotSignal</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_ChIP_H3K27ac_signal</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">2.75</span>, width <span class="op">=</span> <span class="fl">2</span>, height <span class="op">=</span> <span class="fl">0.5</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="co">## Plot GWAS data</span>
<span class="fu"><a href="../../reference/plotManhattan.html">plotManhattan</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">hg19_insulin_GWAS</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    ymax <span class="op">=</span> <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">0.20</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">3.5</span>, width <span class="op">=</span> <span class="fl">2</span>, height <span class="op">=</span> <span class="fl">0.5</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="co">## Plot gene track</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="fu"><a href="../../reference/plotGenes.html">plotGenes</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">4</span>, width <span class="op">=</span> <span class="fl">2</span>, height <span class="op">=</span> <span class="fl">0.5</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="co">## Plot genome label</span>
<span class="fu"><a href="../../reference/plotGenomeLabel.html">plotGenomeLabel</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">4.5</span>, length <span class="op">=</span> <span class="fl">2</span>, scale <span class="op">=</span> <span class="st">"Mb"</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="plotting_multiomic_data_files/figure-html/plotting_example01-1.png" width="750" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="using-the-pgparams-object">Using the <code>pgParams</code> object<a class="anchor" aria-label="anchor" href="#using-the-pgparams-object"></a>
</h2>
<p>The <code><a href="../../reference/pgParams.html">pgParams()</a></code> function creates a <code>pgParams</code> object that can contain any argument from <code>plotgardener</code> functions.</p>
<p>We can recreate and simplify the multi-omic plot above by saving the genomic region, left-based x-coordinate, and width into a <code>pgParams</code> object:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pgParams.html">pgParams</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    width <span class="op">=</span> <span class="fl">2</span>, length <span class="op">=</span> <span class="fl">2</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p>Since these values are the same for each of the functions we are using to build our multi-omic figure, we can now pass the <code>pgParams</code> object into our functions so we don’t need to write the same parameters over and over again:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load example data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_DNAloops_pairs"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_H3K27ac_signal"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"hg19_insulin_GWAS"</span><span class="op">)</span>

<span class="co">## Create a plotgardener page</span>
<span class="fu"><a href="../../reference/pageCreate.html">pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">3</span>, height <span class="op">=</span> <span class="fl">5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>

<span class="co">## Plot Hi-C data in region</span>
<span class="fu"><a href="../../reference/plotHicSquare.html">plotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="fl">0.5</span>, height <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span>

<span class="co">## Plot loop annotations</span>
<span class="fu"><a href="../../reference/plotPairsArches.html">plotPairsArches</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_DNAloops_pairs</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">0.25</span>,
    fill <span class="op">=</span> <span class="st">"black"</span>, linecolor <span class="op">=</span> <span class="st">"black"</span>, flip <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## Plot signal track data</span>
<span class="fu"><a href="../../reference/plotSignal.html">plotSignal</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_ChIP_H3K27ac_signal</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="fl">2.75</span>, height <span class="op">=</span> <span class="fl">0.5</span>
<span class="op">)</span>

<span class="co">## Plot GWAS data</span>
<span class="fu"><a href="../../reference/plotManhattan.html">plotManhattan</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">hg19_insulin_GWAS</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    ymax <span class="op">=</span> <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">0.20</span>,
    y <span class="op">=</span> <span class="fl">3.5</span>, height <span class="op">=</span> <span class="fl">0.5</span>
<span class="op">)</span>

<span class="co">## Plot gene track</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="fu"><a href="../../reference/plotGenes.html">plotGenes</a></span><span class="op">(</span>
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="fl">4</span>, height <span class="op">=</span> <span class="fl">0.5</span>
<span class="op">)</span>

<span class="co">## Plot genome label</span>
<span class="fu"><a href="../../reference/plotGenomeLabel.html">plotGenomeLabel</a></span><span class="op">(</span>
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="fl">4.5</span>, scale <span class="op">=</span> <span class="st">"Mb"</span>
<span class="op">)</span></code></pre></div>
<p><img src="plotting_multiomic_data_files/figure-html/plotting_example02-1.png" width="750" style="display: block; margin: auto;"></p>
<p>The <code>pgParams</code> object also simplifies the code for making complex multi-omic figures when we want to change the genomic region of our plots. If we want to change the region for the figure above, we can simply put it into the <code>pgParams</code> object and re-run the code to generate the figure:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pgParams.html">pgParams</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">29000000</span>, chromend <span class="op">=</span> <span class="fl">30000000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    width <span class="op">=</span> <span class="fl">2</span>, length <span class="op">=</span> <span class="fl">2</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="plotting_multiomic_data_files/figure-html/plotting_example03-1.png" width="750" style="display: block; margin: auto;"></p>
<p>Alternatively, if we want to plot around a particular gene rather than a genomic region we can use <code><a href="../../reference/pgParams.html">pgParams()</a></code> to specify <code>gene</code> and <code>geneBuffer</code>. If <code>geneBuffer</code> is not included, the default buffer adds <code>(gene length) / 2</code> base pairs to the ends of the gene coordinates.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pgParams.html">pgParams</a></span><span class="op">(</span>
    gene <span class="op">=</span> <span class="st">"LINC00113"</span>, geneBuffer <span class="op">=</span> <span class="fl">100000</span>, assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    width <span class="op">=</span> <span class="fl">2</span>, length <span class="op">=</span> <span class="fl">2</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="plotting_multiomic_data_files/figure-html/plotting_example04-1.png" width="750" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="the-below-y-coordinate">The “below” y-coordinate<a class="anchor" aria-label="anchor" href="#the-below-y-coordinate"></a>
</h2>
<p>Since multi-omic plots often involve vertical stacking, the placement of multi-omic plots can be facilitated with the “below” y-coordinate. Rather than providing a <code>numeric</code> value or <code>unit</code> object to the <code>y</code> parameter in plotting functions, we can place a plot below the previously drawn <code>plotgardener</code> plot with a <code>character</code> value consisting of the distance below the last plot, in page units, and “b”. For example, on a page made in inches, <code>y = "0.1b"</code> will place a plot 0.1 inches below the last drawn plot.</p>
<p>We can further simplify the placement code of our multi-omic figure above by using the “below” y-coordinate to easily stack our plots:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load example data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_DNAloops_pairs"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_H3K27ac_signal"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"hg19_insulin_GWAS"</span><span class="op">)</span>

<span class="co">## pgParams</span>
<span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pgParams.html">pgParams</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    width <span class="op">=</span> <span class="fl">2</span>, length <span class="op">=</span> <span class="fl">2</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="co">## Create a plotgardener page</span>
<span class="fu"><a href="../../reference/pageCreate.html">pageCreate</a></span><span class="op">(</span>
    width <span class="op">=</span> <span class="fl">3</span>, height <span class="op">=</span> <span class="fl">5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>,
    showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span>
<span class="op">)</span>

<span class="co">## Plot Hi-C data in region</span>
<span class="fu"><a href="../../reference/plotHicSquare.html">plotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="fl">0.5</span>, height <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span>

<span class="co">## Plot loop annotations</span>
<span class="fu"><a href="../../reference/plotPairsArches.html">plotPairsArches</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_DNAloops_pairs</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="st">"0b"</span>,
    height <span class="op">=</span> <span class="fl">0.25</span>,
    fill <span class="op">=</span> <span class="st">"black"</span>, linecolor <span class="op">=</span> <span class="st">"black"</span>, flip <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## Plot signal track data</span>
<span class="fu"><a href="../../reference/plotSignal.html">plotSignal</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_ChIP_H3K27ac_signal</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="st">"0b"</span>,
    height <span class="op">=</span> <span class="fl">0.5</span>
<span class="op">)</span>

<span class="co">## Plot GWAS data</span>
<span class="fu"><a href="../../reference/plotManhattan.html">plotManhattan</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">hg19_insulin_GWAS</span>,
    params <span class="op">=</span> <span class="va">params</span>,
    ymax <span class="op">=</span> <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">0.20</span>,
    y <span class="op">=</span> <span class="st">"0.25b"</span>,
    height <span class="op">=</span> <span class="fl">0.5</span>
<span class="op">)</span>

<span class="co">## Plot gene track</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>
<span class="fu"><a href="../../reference/plotGenes.html">plotGenes</a></span><span class="op">(</span>
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="st">"0b"</span>,
    height <span class="op">=</span> <span class="fl">0.5</span>
<span class="op">)</span>

<span class="co">## Plot genome label</span>
<span class="fu"><a href="../../reference/plotGenomeLabel.html">plotGenomeLabel</a></span><span class="op">(</span>
    params <span class="op">=</span> <span class="va">params</span>,
    y <span class="op">=</span> <span class="st">"0b"</span>,
    scale <span class="op">=</span> <span class="st">"Mb"</span>
<span class="op">)</span></code></pre></div>
<p><img src="plotting_multiomic_data_files/figure-html/below_y-1.png" width="750" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="plotting-and-comparing-multiple-signal-tracks">Plotting and comparing multiple signal tracks<a class="anchor" aria-label="anchor" href="#plotting-and-comparing-multiple-signal-tracks"></a>
</h2>
<p>In many multi-omic visualizations, multiple signal tracks are often aligned and stacked to compare different kinds of signal data and/or signals from different samples. <strong><code>plotgardener</code> does not normalize signal data</strong> <strong>based on variables like read depth</strong>, but it is possible to <strong>scale</strong> <code>plotgardener</code> signal plots to the same y-axis.</p>
<p>To determine the appropriate y-axis range, we first must get the maximum signal score from all of our datasets to be compared:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/plotgardenerData" class="external-link">plotgardenerData</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_H3K27ac_signal"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GM12878_ChIP_H3K27ac_signal"</span><span class="op">)</span>

<span class="va">maxScore</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="va">IMR90_ChIP_H3K27ac_signal</span><span class="op">$</span><span class="va">score</span>, 
                    <span class="va">GM12878_ChIP_H3K27ac_signal</span><span class="op">$</span><span class="va">score</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">maxScore</span><span class="op">)</span>
<span class="co">#&gt; [1] 40.91454</span></code></pre></div>
<p>In each of our signal plotting calls, we will then use the <code>range</code> parameter to set the range of both our y-axes to <code>c(0, maxScore)</code>. Here we can do this with our <code>pgParams</code> object:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/pgParams.html">pgParams</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>,
    chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    range <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">maxScore</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>We are now ready to plot, align, and compare our signal plots along the genomic x-axis and the score y-axis:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Create a page</span>
<span class="fu"><a href="../../reference/pageCreate.html">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">7.5</span>, height <span class="op">=</span> <span class="fl">2.1</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>,
            showGuides <span class="op">=</span> <span class="cn">FALSE</span>, xgrid <span class="op">=</span> <span class="fl">0</span>, ygrid <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>

<span class="co">## Plot and place signal plots</span>
<span class="va">signal1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotSignal.html">plotSignal</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_ChIP_H3K27ac_signal</span>, params <span class="op">=</span> <span class="va">params</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">6.5</span>, height <span class="op">=</span> <span class="fl">0.65</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="va">signal2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/plotSignal.html">plotSignal</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">GM12878_ChIP_H3K27ac_signal</span>, params <span class="op">=</span> <span class="va">params</span>,
    linecolor <span class="op">=</span> <span class="st">"#7ecdbb"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1</span>, width <span class="op">=</span> <span class="fl">6.5</span>, height <span class="op">=</span> <span class="fl">0.65</span>,
    just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="co">## Plot genome label</span>
<span class="fu"><a href="../../reference/plotGenomeLabel.html">plotGenomeLabel</a></span><span class="op">(</span>
    chrom <span class="op">=</span> <span class="st">"chr21"</span>,
    chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1.68</span>, length <span class="op">=</span> <span class="fl">6.5</span>,
    default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="co">## Add text labels</span>
<span class="fu"><a href="../../reference/plotText.html">plotText</a></span><span class="op">(</span>
    label <span class="op">=</span> <span class="st">"IMR90"</span>, fonsize <span class="op">=</span> <span class="fl">10</span>, fontcolor <span class="op">=</span> <span class="st">"#37a7db"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">0.25</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>
<span class="fu"><a href="../../reference/plotText.html">plotText</a></span><span class="op">(</span>
    label <span class="op">=</span> <span class="st">"GM12878"</span>, fonsize <span class="op">=</span> <span class="fl">10</span>, fontcolor <span class="op">=</span> <span class="st">"#7ecdbb"</span>,
    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1</span>, just <span class="op">=</span> <span class="fu"><a href="../../reference/c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="plotting_multiomic_data_files/figure-html/scale_signal-1.png" width="1125" style="display: block; margin: auto;"></p>
</div>
<div class="section level1">
<h1 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h1>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    grid      stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] org.Hs.eg.db_3.14.0                    </span>
<span class="co">#&gt;  [2] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2</span>
<span class="co">#&gt;  [3] GenomicFeatures_1.46.1                 </span>
<span class="co">#&gt;  [4] AnnotationDbi_1.56.2                   </span>
<span class="co">#&gt;  [5] Biobase_2.54.0                         </span>
<span class="co">#&gt;  [6] GenomicRanges_1.46.1                   </span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.30.0                    </span>
<span class="co">#&gt;  [8] IRanges_2.28.0                         </span>
<span class="co">#&gt;  [9] S4Vectors_0.32.3                       </span>
<span class="co">#&gt; [10] BiocGenerics_0.40.0                    </span>
<span class="co">#&gt; [11] plotgardenerData_1.0.0                 </span>
<span class="co">#&gt; [12] plotgardener_1.1.2                     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-7                matrixStats_0.61.0         </span>
<span class="co">#&gt;  [3] fs_1.5.1                    bit64_4.0.5                </span>
<span class="co">#&gt;  [5] filelock_1.0.2              progress_1.2.2             </span>
<span class="co">#&gt;  [7] RColorBrewer_1.1-2          httr_1.4.2                 </span>
<span class="co">#&gt;  [9] rprojroot_2.0.2             tools_4.1.2                </span>
<span class="co">#&gt; [11] bslib_0.3.1                 utf8_1.2.2                 </span>
<span class="co">#&gt; [13] R6_2.5.1                    DBI_1.1.1                  </span>
<span class="co">#&gt; [15] colorspace_2.0-2            prettyunits_1.1.1          </span>
<span class="co">#&gt; [17] tidyselect_1.1.1            bit_4.0.4                  </span>
<span class="co">#&gt; [19] curl_4.3.2                  compiler_4.1.2             </span>
<span class="co">#&gt; [21] textshaping_0.3.6           xml2_1.3.3                 </span>
<span class="co">#&gt; [23] desc_1.4.0                  DelayedArray_0.20.0        </span>
<span class="co">#&gt; [25] rtracklayer_1.54.0          sass_0.4.0                 </span>
<span class="co">#&gt; [27] scales_1.1.1                rappdirs_0.3.3             </span>
<span class="co">#&gt; [29] pkgdown_2.0.0               systemfonts_1.0.3          </span>
<span class="co">#&gt; [31] stringr_1.4.0               digest_0.6.29              </span>
<span class="co">#&gt; [33] Rsamtools_2.10.0            yulab.utils_0.0.4          </span>
<span class="co">#&gt; [35] rmarkdown_2.11              XVector_0.34.0             </span>
<span class="co">#&gt; [37] pkgconfig_2.0.3             htmltools_0.5.2            </span>
<span class="co">#&gt; [39] MatrixGenerics_1.6.0        highr_0.9                  </span>
<span class="co">#&gt; [41] dbplyr_2.1.1                fastmap_1.1.0              </span>
<span class="co">#&gt; [43] rlang_0.4.12                RSQLite_2.2.8              </span>
<span class="co">#&gt; [45] gridGraphics_0.5-1          jquerylib_0.1.4            </span>
<span class="co">#&gt; [47] BiocIO_1.4.0                generics_0.1.1             </span>
<span class="co">#&gt; [49] jsonlite_1.7.2              BiocParallel_1.28.2        </span>
<span class="co">#&gt; [51] dplyr_1.0.7                 RCurl_1.98-1.5             </span>
<span class="co">#&gt; [53] magrittr_2.0.1              ggplotify_0.1.0            </span>
<span class="co">#&gt; [55] GenomeInfoDbData_1.2.7      Matrix_1.3-4               </span>
<span class="co">#&gt; [57] Rcpp_1.0.7                  munsell_0.5.0              </span>
<span class="co">#&gt; [59] fansi_0.5.0                 lifecycle_1.0.1            </span>
<span class="co">#&gt; [61] stringi_1.7.6               yaml_2.2.1                 </span>
<span class="co">#&gt; [63] SummarizedExperiment_1.24.0 zlibbioc_1.40.0            </span>
<span class="co">#&gt; [65] BiocFileCache_2.2.0         blob_1.2.2                 </span>
<span class="co">#&gt; [67] parallel_4.1.2              crayon_1.4.2               </span>
<span class="co">#&gt; [69] lattice_0.20-45             Biostrings_2.62.0          </span>
<span class="co">#&gt; [71] hms_1.1.1                   KEGGREST_1.34.0            </span>
<span class="co">#&gt; [73] knitr_1.36                  pillar_1.6.4               </span>
<span class="co">#&gt; [75] rjson_0.2.20                biomaRt_2.50.1             </span>
<span class="co">#&gt; [77] strawr_0.0.9                XML_3.99-0.8               </span>
<span class="co">#&gt; [79] glue_1.5.1                  evaluate_0.14              </span>
<span class="co">#&gt; [81] data.table_1.14.2           png_0.1-7                  </span>
<span class="co">#&gt; [83] vctrs_0.3.8                 gtable_0.3.0               </span>
<span class="co">#&gt; [85] purrr_0.3.4                 assertthat_0.2.1           </span>
<span class="co">#&gt; [87] cachem_1.0.6                ggplot2_3.3.5              </span>
<span class="co">#&gt; [89] xfun_0.28                   restfulr_0.0.13            </span>
<span class="co">#&gt; [91] ragg_1.2.0                  tibble_3.1.6               </span>
<span class="co">#&gt; [93] GenomicAlignments_1.30.0    plyranges_1.14.0           </span>
<span class="co">#&gt; [95] memoise_2.0.1               ellipsis_0.3.2</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nekramer" class="external-link">Nicole Kramer</a>, <a href="https://github.com/EricSDavis" class="external-link">Eric S. Davis</a>, <a href="https://github.com/cwenger" class="external-link">Craig Wenger</a>, <a href="https://github.com/dphansti" class="external-link">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.0.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
