---
title: A guide to plot aesthetics
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{Additional plot types and elements}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.align = "center",
    message = FALSE,
    warning = FALSE
)
library(BentoBox)
library(grid)
library(BentoBoxData)
```

`BentoBox` plots are extremely customizable in appearance. In this article 
we will outline some of the major aesthetic customizations, including general 
features and specific plot type customizations.

All the data included in this article can be found in the supplementary
package `BentoBoxData`.

# `gpar` and common plot customizations

The most common types of customizations are inherited from `grid` `gpar` 
options. If a function accepts `...`, this usually refers to `gpar` options 
that are not explicity listed as parameters in the function documentation. 
General valid parameters include:

+------------------:+:--------------------------------------------------------+
| **alpha**         | Alpha channel for transparency (number between 0 and 1).|
+-------------------+---------------------------------------------------------+
| **fill**          | Fill color.                                             |
+-------------------+---------------------------------------------------------+
|  **linecolor**    | Line color.                                             |
+-------------------+---------------------------------------------------------+
| **lty**           | Line type. (0=blank, 1=solid, 2=dashed, 3=dotted,       |
|                   | 4=dotdash, 5=longdash, 6=twodash).                      |
+-------------------+---------------------------------------------------------+
| **lwd**           | Line width.                                             |
+-------------------+---------------------------------------------------------+
| **lineend**       | Line end style (round, butt, square).                   |
+-------------------+---------------------------------------------------------+
| **linejoin**      | Line join style (round, mitre, bevel).                  |
+-------------------+---------------------------------------------------------+
| **linemitre**     | Line mitre limit (number greater than 1).               |
+-------------------+---------------------------------------------------------+
| **fontsize**      | The size of text, in points.                            |
+-------------------+---------------------------------------------------------+
| **fontcolor**     | Text color.                                             |
+-------------------+---------------------------------------------------------+
| **fontface**      | The font face (plain, bold, italic, bold.italic,        |
|                   | oblique).                                               |
+-------------------+---------------------------------------------------------+
| **fontfamily**    | The font family.                                        |
+-------------------+---------------------------------------------------------+
| **cex**           | Scaling multiplier applied to symbols.                  |
+-------------------+---------------------------------------------------------+
| **pch**           | Plotting character, or symbol (integer codes range      |
|                   | from 0 to 25).                                          |
+-------------------+---------------------------------------------------------+

Additional fonts for the `fontfamily` argument can be imported with the 
packages `extrafont` and `showtext`. This makes it possible to incorporate 
special fonts like Times New Roman, Arial, etc. into `BentoBox` figures.

# Backgrounds and baselines

By default, `BentoBox` plots have transparent backgrounds when placed on a 
`bb_page`. In many functions, this background color can be changed with the 
parameter `bg`.

```{r eval=FALSE, echo=TRUE}
bb_plotGenes(
    chrom = "chr8", chromstart = 1000000, chromend = 2000000,
    assembly = "hg19",
    bg = "#f6f6f6",
    x = 0.5, y = 0.5, width = 2, height = 1, just = c("left", "top"),
    default.units = "inches"
)
```


```{r aes_01, eval=TRUE, echo=FALSE, message=FALSE, fig.width=5, fig.height=3}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)
bb_pageCreate(
    width = 3, height = 2, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
genomicRegion <- bb_params(
    chrom = "chr8",
    chromstart = 1000000, chromend = 2000000,
    assembly = "hg19"
)
genesPlot <- bb_plotGenes(
    params = genomicRegion, bg = "#f6f6f6",
    x = 0.5, y = 0.5, width = 2, height = 1,
    just = c("left", "top"), default.units = "inches"
)
```
This makes it easy to clearly see the precise dimensional boundaries of 
`BentoBox` plots.

Some plots also benefit from baselines to quickly show where y = 0. This can 
aid in data interpretation and guide plot annotation placement. Baselines can
be plotted in selective plots with `baseline = TRUE`.

```{r eval=FALSE, echo=TRUE}
bb_plotRanges(
    data = IMR90_ChIP_CTCF_reads,
    chrom = "chr21", chromstart = 29073000, chromend = 29074000,
    assembly = "hg19",
    fill = c("#7ecdbb", "#37a7db"),
    baseline = TRUE, baseline.color = "black",
    x = 0.5, y = 0.25, width = 6.5, height = 4.25,
    just = c("left", "top"), default.units = "inches"
)
```

```{r aes_02, echo=FALSE, fig.height=6, fig.width=9.5}
data("IMR90_ChIP_CTCF_reads")
bb_pageCreate(
    width = 7.5, height = 5, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)

pileupPlot <- bb_plotRanges(
    data = IMR90_ChIP_CTCF_reads,
    chrom = "chr21", chromstart = 29073000, chromend = 29074000,
    assembly = "hg19",
    fill = c("#7ecdbb", "#37a7db"),
    x = 0.5, y = 0.25, width = 6.5, height = 4.25,
    baseline = TRUE, baseline.color = "black",
    just = c("left", "top"), default.units = "inches"
)
```

# `colorby`

The `colorby` constructor allows us to color the data elements in `BentoBox` 
plots by various data features. These features can be a numerical range, like 
some kind of score value, or categorical values, like positive or negative 
strand. The `colorby` object is constructed by specifying the name of the data 
column to color by, an optional color palette function, and an optional range 
for numerical values. If not specified, `BentoBox` will use the `RColorBrewer`
"YlGnBl" palette for mapping numerical data and the "Pairs" palette for 
qualitative data.

For example, if we revist the BED plot above, `IMR90_ChIP_CTCF_reads` has an additional 
`strand` column for each BED element: 

```{r eval = TRUE, echo = TRUE}
data("IMR90_ChIP_CTCF_reads")
head(IMR90_ChIP_CTCF_reads)
```

Thus, we can use the `colorby` constructor to color BED elements by positive 
or negative strand. The `strand` column will be converted to a factor with a 
`-` level and `+` level. These values will be mapped to our input `palette`:

```{r eval=FALSE, echo=TRUE}
set.seed(nrow(IMR90_ChIP_CTCF_reads))
bb_plotRanges(
    data = IMR90_ChIP_CTCF_reads,
    chrom = "chr21", chromstart = 29073000, chromend = 29074000,
    assembly = "hg19",
    fill = colorby("strand", palette = 
                       colorRampPalette(c("#7ecdbb", "#37a7db"))),
    x = 0.5, y = 0.25, width = 6.5, height = 4.25,
    just = c("left", "top"), default.units = "inches"
)
```

```{r aes_03, echo=FALSE, fig.height=6, fig.width=9.5}

bb_pageCreate(
    width = 7.5, height = 5, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)

set.seed(nrow(IMR90_ChIP_CTCF_reads))
bb_plotRanges(
    data = IMR90_ChIP_CTCF_reads,
    chrom = "chr21", chromstart = 29073000, chromend = 29074000,
    assembly = "hg19",
    fill = colorby("strand", palette = 
                       colorRampPalette(c("#7ecdbb", "#37a7db"))),
    strandSplit = TRUE,
    x = 0.5, y = 0.25, width = 6.5, height = 4.25,
    just = c("left", "top"), default.units = "inches"
)
```

To further control the order of color mapping, we can set our categorical 
`colorby` column as a factor with our own order of levels before plotting:

```{r eval=TRUE, echo=TRUE}
data("IMR90_ChIP_CTCF_reads")
IMR90_ChIP_CTCF_reads$strand <- factor(IMR90_ChIP_CTCF_reads$strand, levels = c("+", "-"))
head(IMR90_ChIP_CTCF_reads$strand)
```

Now we've set the `+` level as our first level, so our `palette` will map
colors in the opposite order from before:

```{r aes_04, echo=FALSE, fig.height=6, fig.width=9.5}

bb_pageCreate(
    width = 7.5, height = 5, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
set.seed(nrow(IMR90_ChIP_CTCF_reads))
bb_plotRanges(
    data = IMR90_ChIP_CTCF_reads,
    chrom = "chr21", chromstart = 29073000, chromend = 29074000,
    assembly = "hg19",
    fill = colorby("strand", palette = 
                       colorRampPalette(c("#7ecdbb", "#37a7db"))),
    strandSplit = TRUE,
    x = 0.5, y = 0.25, width = 6.5, height = 4.25,
    just = c("left", "top"), default.units = "inches"
)
```

In this example, we will color BEDPE arches by a range of numerical values 
we will add as a `length` column:

```{r eval=TRUE, echo=TRUE }
data("IMR90_DNAloops_pairs")
IMR90_DNAloops_pairs$length <- (IMR90_DNAloops_pairs$start2 - IMR90_DNAloops_pairs$start1) / 1000
head(IMR90_DNAloops_pairs$length)
```

Now we can set `fill` as a `colorby` object to color the BEDPE `length` column by:
```{r eval=FALSE, echo=TRUE}
bedpePlot <- bb_plotPairsArches(
    data = IMR90_DNAloops_pairs,
    chrom = "chr21", chromstart = 27900000, chromend = 30700000,
    assembly = "hg19",
    fill = colorby("length", 
                   palette = colorRampPalette(c("dodgerblue2", "firebrick2"))),
    linecolor = "fill",
    archHeight = IMR90_DNAloops_pairs$length / max(IMR90_DNAloops_pairs$length),
    alpha = 1,
    x = 0.25, y = 0.25, width = 7, height = 1.5,
    just = c("left", "top"),
    default.units = "inches"
)
```

```{r aes_05, echo=FALSE, fig.width=9.5, message=FALSE, height=3}

bb_pageCreate(
    width = 7.5, height = 2, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
heights <- IMR90_DNAloops_pairs$length / max(IMR90_DNAloops_pairs$length)
bedpePlot <- bb_plotPairsArches(
    data = IMR90_DNAloops_pairs,
    chrom = "chr21", chromstart = 27900000, chromend = 30700000,
    assembly = "hg19",
    fill = colorby("length", 
                   palette = colorRampPalette(c("dodgerblue2", "firebrick2"))),
    linecolor = "fill",
    archHeight = heights, alpha = 1,
    x = 0.25, y = 0.25, width = 7, height = 1.5,
    just = c("left", "top"),
    default.units = "inches"
)
```

And now since we have numbers mapped to colors, we can use 
`bb_annoHeatmapLegend()` with our `bb_arches` object to add a legend for 
the `colorby` we performed:

```{r eval=FALSE, echo=TRUE}
bb_annoHeatmapLegend(
    plot = bedpePlot, fontcolor = "black",
    x = 7.0, y = 0.25,
    width = 0.10, height = 1, fontsize = 10
)
```
```{r aes_06, echo=FALSE, fig.width=9.5, message=FALSE, height=3}

bb_pageCreate(
    width = 7.5, height = 2, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
heights <- IMR90_DNAloops_pairs$length / max(IMR90_DNAloops_pairs$length)
bedpePlot <- bb_plotPairsArches(
    data = IMR90_DNAloops_pairs,
    chrom = "chr21", chromstart = 27900000, chromend = 30700000,
    assembly = "hg19",
    fill = colorby("length", 
                   palette = colorRampPalette(c("dodgerblue2", "firebrick2"))),
    linecolor = "fill",
    archHeight = heights, alpha = 1,
    x = 0.25, y = 0.25, width = 7, height = 1.5,
    just = c("left", "top"),
    default.units = "inches"
)
bb_annoHeatmapLegend(
    plot = bedpePlot, fontcolor = "black",
    x = 7.0, y = 0.25,
    width = 0.10, height = 1, fontsize = 10
)
```

# Gene and transcript plot aesthetics

`BentoBox` provides many useful features specific for enhancing gene and 
transcript visualizations:

## Labels

Since `BentoBox` utilizes `TxDb` objects, `orgDb` objects, and internal 
citation information, `BentoBox` has access to numerous gene and transcript 
identifiers and can customize annotation labels in a variety of ways. 

By default, `BentoBox` will rank gene labels according to citation number to 
prevent label overcrowding. However, we can provide our own list of 
prioritized genes to label in a plot. For example, if we plot the hg19 genes 
in the region `chr2:1000000-20000000`, our plot will show these labels:

```{r aes_07, eval=TRUE, echo=TRUE, message=FALSE, fig.width=5, fig.height=1.5}
bb_pageCreate(
    width = 5, height = 1.25,
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
genePlot <- bb_plotGenes(
    chrom = "chr2", chromstart = 1000000, chromend = 20000000,
    assembly = "hg19",
    x = 0.25, y = 0.25, width = 4.75, height = 1
)
```
Looking in the `bb_genes` object, we can see that there were numerous genes 
that were not labeled.

```{r eval=TRUE, echo=TRUE}
genePlot$genes
```

If we were particularly interested in MIR3125, we could include this in 
the `geneOrder` parameter to prioritize its labeling:
```{r aes_08, echo=TRUE, fig.height=1.5, fig.width=5, message=FALSE}
bb_pageCreate(
    width = 5, height = 1.25,
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
genePlot <- bb_plotGenes(
    chrom = "chr2", chromstart = 1000000, chromend = 20000000,
    assembly = "hg19",
    geneOrder = c("MIR3125"),
    x = 0.25, y = 0.25, width = 4.75, height = 1
)
```

If we wanted to change the type of label in our plot, we can change 
the `display.column` parameter in a `bb_assembly()` object. The default 
`display.column = "SYMBOL"`, but we could change this value to other 
available `keytypes` in the `orgDb` we are using. For example, if we wanted
to display the associated Ensembl IDs, we would set 
`display.column = "ENSEMBL"`:

```{r aes_09, echo=TRUE, fig.height=2.25, fig.width=5, message=FALSE}
new_hg19 <- bb_assembly(
    Genome = "id_hg19",
    TxDb = "TxDb.Hsapiens.UCSC.hg19.knownGene",
    OrgDb = "org.Hs.eg.db",
    gene.id.column = "ENTREZID",
    display.column = "ENSEMBL"
)
bb_pageCreate(
    width = 5, height = 1.25,
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
genePlot <- bb_plotGenes(
    chrom = "chr2", chromstart = 1000000, chromend = 20000000,
    assembly = new_hg19,
    x = 0.25, y = 0.25, width = 4.75, height = 1
)
```

Label IDs used in transcript plots can be customized through `bb_assembly()` 
objects, and transcript label formatting can be changed through the `labels` 
parameter. For example, if we wish to display both transcript names and their 
associated gene names, we can set `labels = "both"`:

```{r aes_10, echo=TRUE, fig.height=6, fig.width=7}
bb_pageCreate(
    width = 6, height = 5,
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
transcriptPlot <- bb_plotTranscripts(
    chrom = "chr2", chromstart = 1000000, chromend = 20000000,
    assembly = "hg19",
    labels = "both",
    x = 0.25, y = 0.25, width = 5.5, height = 4.5
)
```

## Highlighting genes by color

In addition to changing `fill` and `fontcolor` to change the colors of all 
genes in a plot, specific gene structures and their labels can be highlighted
in a different color with `geneHighlights`. If we revisit the `bb_genes` plot
above, we can highlight RRM2 by creating a `data.frame` with "RRM2" in the 
first column and its highlight color in the second column:

```{r eval=TRUE, echo=TRUE, message=FALSE}
geneHighlights <- data.frame("geneName" = "RRM2", "color" = "steel blue")
```

We can then pass this into our `bb_plotGenes()` call:

```{r aes_11, eval=TRUE, echo=TRUE, message=FALSE, fig.width=5, fig.height=1.5}
bb_pageCreate(
    width = 5, height = 1.25,
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
genePlot <- bb_plotGenes(
    chrom = "chr2", chromstart = 1000000, chromend = 20000000,
    assembly = "hg19",
    geneHighlights = geneHighlights, geneBackground = "grey",
    x = 0.25, y = 0.25, width = 4.75, height = 1
)
```

Since `geneHighlights` is a `data.frame`, we can highlight multiple genes 
in different colors at once. For example, let's now highlight RRM2 in 
"steel blue" and PXDN in "red":

```{r aes_12, eval=TRUE, echo=TRUE, message=FALSE, fig.width=5, fig.height=1.5}
geneHighlights <- data.frame(
    "geneName" = c("RRM2", "PXDN"),
    "color" = c("steel blue", "red")
)

bb_pageCreate(
    width = 5, height = 1.25,
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
genePlot <- bb_plotGenes(
    chrom = "chr2", chromstart = 1000000, chromend = 20000000,
    assembly = "hg19",
    geneHighlights = geneHighlights, geneBackground = "grey",
    x = 0.25, y = 0.25, width = 4.75, height = 1
)
```

## Customizing transcripts by strand

To distinguish which strand a transcript belongs to, `bb_plotTranscripts()` 
colors transcripts by strand with the parameter `colorbyStrand`. The first 
value in `fill` colors positive strand transcripts and the second `fill`
value colors negative strand transcripts. To further organize transcripts by 
strand, we can use `strandSplit` to separate transcript elements into groups 
of positive and negative strands:

```{r aes_13, echo=TRUE, fig.height=6, fig.width=7}
bb_pageCreate(
    width = 6, height = 5,
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
transcriptPlot <- bb_plotTranscripts(
    chrom = "chr2", chromstart = 1000000, chromend = 20000000,
    assembly = "hg19",
    strandSplit = TRUE,
    x = 0.25, y = 0.25, width = 5.5, height = 4.5
)
```

Now all our positive strand transcripts are grouped together above the 
group of negative strand transcripts.

# Hi-C plot customizations

`BentoBox` includes many types of customizations for Hi-C plots. `BentoBox` 
provides 3 different Hi-C plotting functions based on the desired plot shape:

- `bb_plotHicSquare()`: Plots a square, symmetrical Hi-C plot with genomic 
coordinates along both the x- and y-axes.
    
- `bb_plotHicTriangle()`: Plots a triangular Hi-C plot where the genome 
region falls along the base of the triangle.
    
- `bb_plotHicRectangle()`: Plots a triangular Hi-C plot with additional data 
filling the surrounding regions to form a rectangle.

```{r aes_14, echo=FALSE, fig.height=3.9, fig.width=9.75, message=FALSE}
data("IMR90_HiC_10kb")

bb_pageCreate(
    width = 7.1, height = 2.25, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
params <- bb_params(
    chrom = "chr21", chromstart = 28950000, chromend = 29800000,
    assembly = "hg19", resolution = 10000
)

hicPlot_square <- bb_plotHicSquare(
    data = IMR90_HiC_10kb, params = params,
    zrange = c(0, 70),
    x = 0.25, y = 0.25,
    width = 1.6, height = 1.6
)
bb_annoGenomeLabel(
    plot = hicPlot_square, fontsize = 8,
    x = 0.25, y = 1.9, scale = "Kb"
)
bb_annoGenomeLabel(
    plot = hicPlot_square, axis = "y", fontsize = 8,
    x = 0.20, y = 0.25, scale = "Kb",
    just = c("right", "top")
)
hicPlot_triangle <- bb_plotHicTriangle(
    data = IMR90_HiC_10kb, params = params,
    zrange = c(0, 70),
    x = 2.1, y = 1.85,
    width = 2.25, height = 2,
    just = c("left", "bottom")
)
bb_annoGenomeLabel(
    plot = hicPlot_triangle, fontsize = 8,
    x = 2.1, y = 1.9, scale = "Kb"
)
hicPlot_rectangle <- bb_plotHicRectangle(
    data = IMR90_HiC_10kb, params = params,
    zrange = c(0, 70),
    x = 4.6, y = 1.85,
    width = 2.25, height = 1.125,
    just = c("left", "bottom")
)
bb_annoGenomeLabel(
    plot = hicPlot_rectangle, fontsize = 8,
    x = 4.6, y = 1.9, scale = "Kb"
)
```  

All Hi-C plot types can use different color palettes, and colors can be 
linearly or log-scaled with the `colorTrans` parameter.  

`bb_hicSquare` plots can be further customized to include two datasets in 
one plot. Instead of plotting both symmetrical halves of the plot, we can set 
one dataset as `half = "top"` and the other dataset as `half = "bottom"`:

```{r aes_15, echo=TRUE, fig.height=4.25, fig.width=5.25, message=FALSE}
data("GM12878_HiC_10kb")
data("IMR90_HiC_10kb")

bb_pageCreate(
    width = 3.25, height = 3.25, default.units = "inches",
    showGuides = FALSE, xgrid = 0, ygrid = 0
)
params <- bb_params(
    chrom = "chr21", chromstart = 28000000, chromend = 30300000,
    assembly = "hg19", resolution = 10000,
    x = 0.25, width = 2.75, just = c("left", "top"), default.units = "inches"
)


hicPlot_top <- bb_plotHicSquare(
    data = GM12878_HiC_10kb, params = params,
    zrange = c(0, 200),
    half = "top",
    y = 0.25, height = 2.75
)
hicPlot_bottom <- bb_plotHicSquare(
    data = IMR90_HiC_10kb, params = params,
    zrange = c(0, 70),
    half = "bottom",
    y = 0.25, height = 2.75
)
```
