<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • plotgardener</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"><meta property="og:image" content="https://phanstiellab.github.io/plotgardener/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plotgardener</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/guides/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../articles/introduction_to_plotgardener.html">Introduction to plotgardener</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/guides/plotgardener_page.html">The plotgardener Page</a>
    </li>
    <li>
      <a href="../articles/guides/reading_data_for_plotgardener.html">Reading Data for plotgardener</a>
    </li>
    <li>
      <a href="../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../articles/guides/plotgardener_meta_functions.html">plotgardener Meta Functions</a>
    </li>
    <li>
      <a href="../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../articles/guides/incorporating_ggplots.html">Incorporating ggplots and other grid-based Bioconductor visualizations</a>
    </li>
    <li>
      <a href="../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/PhanstielLab/plotgardener" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/PhanstielLab/plotgardener/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.2" id="changes-in-version-122">CHANGES IN VERSION 1.2.2<a class="anchor" aria-label="anchor" href="#changes-in-version-122"></a></h2>
<ul><li>Note about double page rendering has been added to <code><a href="../reference/pageGuideHide.html">pageGuideHide()</a></code> documentation.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.1" id="changes-in-version-121">CHANGES IN VERSION 1.2.1<a class="anchor" aria-label="anchor" href="#changes-in-version-121"></a></h2>
<ul><li><p><code>plotSignal</code> bug fixes related to function not finding posSignal2 and negSignal2 variables with insufficient data.</p></li>
<li><p>Documentation to introduction vignette has been added to explain double page rendering when using any removal function, particularly <code><a href="../reference/pageGuideHide.html">pageGuideHide()</a></code>.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.0" id="version-120">VERSION 1.2.0<a class="anchor" aria-label="anchor" href="#version-120"></a></h2>
<p>Version bump for Bioconductor 3.15 release.</p>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.18" id="changes-in-version-1118">CHANGES IN VERSION 1.1.18<a class="anchor" aria-label="anchor" href="#changes-in-version-1118"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-18">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-1-1-18"></a></h3>
<ul><li>
<code>hicTriangles</code> and <code>hicRectangles</code> can now be annotated with <code>annoDomains</code> or <code>annoPixels</code> if they are flipped.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.17" id="changes-in-version-1117">CHANGES IN VERSION 1.1.17<a class="anchor" aria-label="anchor" href="#changes-in-version-1117"></a></h2>
<div class="section level3">
<h3 id="new-features-1-1-17">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-1-1-17"></a></h3>
<ul><li>
<code>plotIdeogram</code> can now accept custom colors with a <code>fill</code> parameter. Colors can be specified with a named or unnamed vector. To see which stains are being assigned which colors, look inside the <code>ideogram</code> object.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.16" id="changes-in-version-1116">CHANGES IN VERSION 1.1.16<a class="anchor" aria-label="anchor" href="#changes-in-version-1116"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-16">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-16"></a></h3>
<ul><li>ENTREZ IDs obtained from <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">AnnotationDbi::select()</a></code> are subset just for <code>ENTREZID</code> column when determining default gene priorities, eliminating <code>dplyr</code> incompatible types error.</li>
<li>All plus and minus strand gene name label parsing in <code>plotGenes</code> is now carried out only if there is a non-zero number of that strand’s genes.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.15" id="changes-in-version-1115">CHANGES IN VERSION 1.1.15<a class="anchor" aria-label="anchor" href="#changes-in-version-1115"></a></h2>
<ul><li>Citation linked for <code>plotgardener</code> publication in Bioinformatics.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.14" id="changes-in-version-1114">CHANGES IN VERSION 1.1.14<a class="anchor" aria-label="anchor" href="#changes-in-version-1114"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-14">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-14"></a></h3>
<ul><li>
<code>plotSignal</code> yrange parsing for negative scores now has fixed the typo on line 418 from “score2” to “score”.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.13" id="changes-in-version-1113">CHANGES IN VERSION 1.1.13<a class="anchor" aria-label="anchor" href="#changes-in-version-1113"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-13">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-13"></a></h3>
<ul><li>
<code>plotSignal</code> default yrange parsing now catches the invalid 0,0 range and no longer throws a viewport related error.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.12" id="changes-in-version-1112">CHANGES IN VERSION 1.1.12<a class="anchor" aria-label="anchor" href="#changes-in-version-1112"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-12">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-12"></a></h3>
<ul><li>
<code>readHic</code> and functions related to the reading of .hic files now leaves the chromosome input formatted as is (e.g. “chr1” and “1”). Functions will throw an error if the input chromosome is not found in the chromosomes listed in the .hic file.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.1.11" id="changes-in-version-1111">CHANGES IN VERSION 1.1.11<a class="anchor" aria-label="anchor" href="#changes-in-version-1111"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-1-1-11">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-1-1-11"></a></h3>
<ul><li>
<code>annoDomains</code> coordinates fixed for <code>plotHicRectangle</code>.</li>
<li>Clipping logic for <code>plotPairsArches</code> now clips arches both on left and right side of plot.</li>
<li>Subsetting <code>plotPairs</code> logic fixed to match <code>plotPairsArches</code>.</li>
</ul></div>
<div class="section level3">
<h3 id="new-features-1-1-11">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-1-1-11"></a></h3>
<ul><li>
<code>clip.noAnchors</code> parameter in <code>plotPairsArches</code> allows for inclusion or clipping of arches that do not have anchors in the given genomic region.</li>
<li>
<code>plotPairsArches</code> now allows for column name input to designate <code>archHeights</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.11" id="changes-in-version-09911">CHANGES IN VERSION 0.99.11<a class="anchor" aria-label="anchor" href="#changes-in-version-09911"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-99-11">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-0-99-11"></a></h3>
<ul><li>
<code>:</code> added back to readHic for <code>strawr</code> region parsing.</li>
<li>
<code>plotHicSquare</code> subsetting fixed for off diagonal regions.</li>
</ul></div>
<div class="section level3">
<h3 id="new-features-0-99-11">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-0-99-11"></a></h3>
<ul><li>All Hi-C functions now allow input of remote Hi-C files.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.9" id="changes-in-version-0999">CHANGES IN VERSION 0.99.9<a class="anchor" aria-label="anchor" href="#changes-in-version-0999"></a></h2>
<p>This package was previously called <code>BentoBox</code>.</p>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.0" id="changes-in-version-0990">CHANGES IN VERSION 0.99.0<a class="anchor" aria-label="anchor" href="#changes-in-version-0990"></a></h2>
<div class="section level3">
<h3 id="new-features-0-99-0">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-0-99-0"></a></h3>
<ul><li>Version bump to 0.99.0 for Bioconductor package submission.</li>
<li>
<code>bb_mapColors</code> function for users to map a vector to a palette of colors.</li>
<li>
<code>linecolor</code> parameter in <code>bb_plotPairs</code>, <code>bb_plotPairsArches</code>, and <code>bb_plotRanges</code> now accepts a single value, a vector of colors, a <code>colorby</code> object, or the value “fill”.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.0.14" id="changes-in-version-00014">CHANGES IN VERSION 0.0.0.14<a class="anchor" aria-label="anchor" href="#changes-in-version-00014"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-0-14">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-0-14"></a></h3>
<ul><li>R version requirement changed to (R &gt;= 4.1.0) for proper plot placement.</li>
</ul></div>
<div class="section level3">
<h3 id="new-features-0-0-0-14">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-0-0-0-14"></a></h3>
<ul><li>
<code>colorby</code> object now has a <code>scalePerRegion</code> parameter to scale numerical <code>colorby</code> data to the range of data in a plotted genomic region.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.0.13" id="changes-in-version-00013">CHANGES IN VERSION 0.0.0.13<a class="anchor" aria-label="anchor" href="#changes-in-version-00013"></a></h2>
<div class="section level3">
<h3 id="significant-user-visible-changes-0-0-0-13">SIGNIFICANT USER-VISIBLE CHANGES<a class="anchor" aria-label="anchor" href="#significant-user-visible-changes-0-0-0-13"></a></h3>
<ul><li>
<code>bb_plotManhattan</code> <code>fill</code> paramete now accepts a single value, a vector of colors, or a <code>colorby</code> object.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.0.12" id="changes-in-version-00012">CHANGES IN VERSION 0.0.0.12<a class="anchor" aria-label="anchor" href="#changes-in-version-00012"></a></h2>
<div class="section level3">
<h3 id="significant-user-visible-changes-0-0-0-12">SIGNIFICANT USER-VISIBLE CHANGES<a class="anchor" aria-label="anchor" href="#significant-user-visible-changes-0-0-0-12"></a></h3>
<ul><li>
<code>colorby</code> constructor now includes optional palette specification.</li>
<li>
<code>bb_plotPairs</code>, <code>bb_plotPairsArches</code>, and <code>bb_plotRanges</code> <code>fill</code> parameter now accepts a single value, a vector of colors, or a <code>colorby</code> object.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-0-12">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-0-12"></a></h3>
<ul><li>
<code>GInteractions</code> assembly match checking moved before dataframe conversion.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.0.11" id="changes-in-version-00011">CHANGES IN VERSION 0.0.0.11<a class="anchor" aria-label="anchor" href="#changes-in-version-00011"></a></h2>
<div class="section level3">
<h3 id="significant-user-visible-changes-0-0-0-11">SIGNIFICANT USER-VISIBLE CHANGES<a class="anchor" aria-label="anchor" href="#significant-user-visible-changes-0-0-0-11"></a></h3>
<ul><li>Data moved to <code>plotgardenerData</code> package.</li>
<li>Default genome assembly updated to “hg38”.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-0-11">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-0-11"></a></h3>
<ul><li>Streamlined parameter parsing and data reading logic.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.0.10" id="changes-in-version-00010">CHANGES IN VERSION 0.0.0.10<a class="anchor" aria-label="anchor" href="#changes-in-version-00010"></a></h2>
<div class="section level3">
<h3 id="new-features-0-0-0-10">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-0-0-0-10"></a></h3>
<ul><li>Added unit tests with <code>testthat</code>.</li>
<li>
<code>bb_annoDomains</code> function addition.</li>
<li>
<code>bb_plotSignal</code> vertical orientation.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.0.9" id="changes-in-version-0009">CHANGES IN VERSION 0.0.0.9<a class="anchor" aria-label="anchor" href="#changes-in-version-0009"></a></h2>
<div class="section level3">
<h3 id="new-features-0-0-0-9">NEW FEATURES<a class="anchor" aria-label="anchor" href="#new-features-0-0-0-9"></a></h3>
<ul><li>Added a <code>NEWS</code> file to track changes to the package.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-0-9">BUG FIXES<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-0-9"></a></h3>
<ul><li>Updated viewport parsing for package <code>grid</code> version 4.1.0.</li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/nekramer" class="external-link">Nicole Kramer</a>, <a href="https://github.com/EricSDavis" class="external-link">Eric S. Davis</a>, <a href="https://github.com/cwenger" class="external-link">Craig Wenger</a>, <a href="https://github.com/dphansti" class="external-link">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

