---
title: Our Philosophy
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Our Philosophy}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.height = 5,
  fig.width = 6,
  fig.align = "center",
  fig.path = "man/figures/"
)
library(BentoBox)
```

**`BentoBox` was developed to give users unparalleled control in plotting and arranging multiple figures entirely within the R plotting environment. This functionality is motivated by the following design principles:**

## Programmatic and reproducible multi-panel figures

Data visualization is most rigorous and accurate when it is reproducible. Although many visualizations are initially designed with code, final customizations and arrangements of figures are typically made with third party graphic design software and cannot be easily replicated. `BentoBox` makes this entire process programmatic, from raw data input to finalized, complex visualization. This means that if users define the same data:

```{r eval=TRUE, echo = TRUE, message=FALSE,warning=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)
data("bb_hicData")
head(bb_hicData)
```

and the same `BentoBox` code:

```{r eval=FALSE}
## Create a BentoBox page
bb_pageCreate(width = 4, height = 3.5, default.units = "inches")

## Define genomic region with `bb_params`
params <- bb_params(chrom = "chr21", chromstart = 28000000, chromend = 30300000)

## Plot Hi-C data
hicPlot <- bb_plotHicTriangle(data = bb_hicData,
                              params = params,
                              x = 2, y = 0.5, width = 3, height = 2, 
                              just = "top", default.units = "inches")

## Add color scale annotation
bb_annoHeatmapLegend(plot = hicPlot, 
                     x = 3, y = 0.5, width = 0.12, height = 0.6, 
                     just = c("left", "top"), default.units = "inches")

## Plot gene track in same genomic region
genesPlot <- bb_plotGenes(params = params,
                          x = 0.5, y = 2.25, width = 3, height = 0.5,
                          just = c("left", "top"), default.units = "inches")

## Label genomic region
bb_annoGenomeLabel(plot = genesPlot,
                   x = 0.5, y = 2.8, just = c("left", "top"), default.units = "inches")

## Hide page guides
bb_pageGuideHide()

```

they will generate the **exact same** visualization:

```{r philosophy_1, eval=TRUE, message=FALSE, echo=FALSE}
## Create a BentoBox page
bb_pageCreate(width = 4, height = 3.5, default.units = "inches",
              xgrid = 0, ygrid = 0, showOutline = FALSE,
              showRuler = FALSE, showUnit = FALSE)

## Define genomic region with `bb_params`
params <- bb_params(chrom = "chr21", chromstart = 28000000, chromend = 30300000)

## Plot Hi-C data
hicPlot <- bb_plotHicTriangle(data = bb_hicData,
                              params = params,
                              x = 2, y = 0.5, width = 3, height = 2, 
                              just = "top", default.units = "inches")

## Add color scale annotation
bb_annoHeatmapLegend(plot = hicPlot, 
                     x = 3, y = 0.5, width = 0.12, height = 0.6, 
                     just = c("left", "top"), default.units = "inches")

## Plot gene track in same genomic region
genesPlot <- bb_plotGenes(params = params,
                          x = 0.5, y = 2.25, width = 3, height = 0.5,
                          just = c("left", "top"), default.units = "inches")

## Label genomic region
bb_annoGenomeLabel(plot = genesPlot,
                   x = 0.5, y = 2.8, just = c("left", "top"), default.units = "inches")

```

Not only does this make visualizations entirely reproducible with code, but it also makes it easy to change data ranges and aesthetics with simple pieces of code:

```{r eval=FALSE}
## Change colors while maintaining plot layout
hicPlot <- bb_plotHicTriangle(data = bb_hicData,
                              params = params,
                              palette = colorRampPalette(c("white", "steel blue")),
                              x = 2, y = 0.5, width = 3, height = 2, 
                              just = "top", default.units = "inches")

genesPlot <- bb_plotGenes(params = params,
                          strandcolors = c("grey", "grey"), fontcolors = c("black", "black"),
                          x = 0.5, y = 2.25, width = 3, height = 0.5,
                          just = c("left", "top"), default.units = "inches")
```

```{r philosophy_2, eval=TRUE, message=FALSE, echo=FALSE}
## Create a BentoBox page
bb_pageCreate(width = 4, height = 3.5, default.units = "inches",
              xgrid = 0, ygrid = 0, showOutline = FALSE,
              showRuler = FALSE, showUnit = FALSE)

## Define genomic region with `bb_params`
params <- bb_params(chrom = "chr21", chromstart = 28000000, chromend = 30300000)

## Plot Hi-C data
hicPlot <- bb_plotHicTriangle(data = bb_hicData,
                              params = params,
                              palette = colorRampPalette(c("white", "steel blue")),
                              x = 2, y = 0.5, width = 3, height = 2, 
                              just = "top", default.units = "inches")

## Add color scale annotation
bb_annoHeatmapLegend(plot = hicPlot, 
                     x = 3, y = 0.5, width = 0.12, height = 0.6, 
                     just = c("left", "top"), default.units = "inches")

## Plot gene track in same genomic region
genesPlot <- bb_plotGenes(params = params,
                          strandcolors = c("grey", "grey"), fontcolors = c("black", "black"),
                          x = 0.5, y = 2.25, width = 3, height = 0.5,
                          just = c("left", "top"), default.units = "inches")

## Label genomic region
bb_annoGenomeLabel(plot = genesPlot,
                   x = 0.5, y = 2.8, just = c("left", "top"), default.units = "inches")

```

## Coordinate-based plot placement

A coordinate-based plotting system provides users with exquisite control over plot sizes and arrangements through a user-defined page and common units of measurement. This system makes the `BentoBox` plotting process intuitive and absolute, meaning that plots cannot be squished and stretched based on their *relative* sizes and placements. Users can create a page in their preferred size and unit of measurement:

```{r philosophy_3, eval=TRUE}
bb_pageCreate(width = 3, height = 3, default.units = "inches")
```

and then make and precisely arrange plots within this defined landscape. If users want the top, left corner of their plot to be 0.5 inches down from the top of this page and 0.5 inches from the left:

```{r philosphy_4, echo=FALSE, eval=TRUE, message=FALSE}
bb_pageCreate(width = 3, height = 3, default.units = "inches")
bb_plotSegments(x0 = 0.5, y0 = 0, x1 = 0.5, y1 = 0.5, arrow = arrow(angle=30,length=unit(0.1, "inches"),ends="last",type="closed"), fill = "black")
bb_plotSegments(x0 = 0, y0 = 0.5, x1 = 0.5, y1 = 0.5, arrow = arrow(angle=30,length=unit(0.1, "inches"),ends="last",type="closed"), fill = "black")
bb_plotText(label = "0.5 in", x =0.6, y = 0.25, just = "left")
bb_plotText(label = "0.5 in", x = 0.25, y = 0.6, just = "top")

```

and users want their plot be 2 inches wide and 1 inch tall:

```{r philosphy_5, echo=FALSE, eval=TRUE, message=FALSE}
bb_pageCreate(width = 3, height = 3, default.units = "inches")
bb_plotSegments(x0 = 0.5, y0 = 0, x1 = 0.5, y1 = 0.5, arrow = arrow(angle=30,length=unit(0.1, "inches"),ends="last",type="closed"), fill = "black")
bb_plotSegments(x0 = 0, y0 = 0.5, x1 = 0.5, y1 = 0.5, arrow = arrow(angle=30,length=unit(0.1, "inches"),ends="last",type="closed"), fill = "black")
bb_plotSegments(x0 = 0.5 , y0 = 0.5, x1 = 2.5, y1 = 0.5, fill = "black", arrow = arrow(angle=30, length=unit(0.1,"inches"),ends="both",type="closed"))
bb_plotSegments(x0 = 0.5 , y0 = 0.5, x1 = 0.5, y1 = 1.5, fill = "black", arrow = arrow(angle=30, length=unit(0.1,"inches"),ends="both",type="closed"))
bb_plotText(label = "2 in", x = 1.5, y = 0.4, just = "bottom")
bb_plotText(label = "1 in", x = 0.4, y = 1, just = "right")

```

`BentoBox` can make the plot with these exact parameters:
```{r philosophy_6, echo=TRUE, eval=TRUE, message=FALSE}
## Create BentoBox page
bb_pageCreate(width = 3, height = 3, default.units = "inches")

## Load signal data
data("bb_signalData")

## Plot and place signal data with precise measurements
signalPlot <- bb_plotSignal(data = bb_signalData,
                            chrom = "chr21", chromstart = 28000000, chromend = 30300000,
                            x = 0.5, y = 0.5, width = 2, height = 1, 
                            just = c("left", "top"), default.units = "inches")

```

This exact plot sizing and placement is particularly important and useful for standardized and accurate data comparisons along the x and y-axes. 

## Containerized, edge-to-edge visualizations

With the stacked nature of genomic plots, it is critical that aligned plots correspond to the same genomic regions to ensure figure accuracy. By separating data plotting from plot annotations, data fills defined plot coordinates and dimensions from edge to edge and precisely preserves the mapping between the data and its user-defined container. For example, if a user creates a 2 inch-wide gene track and wants that plot to represent the genomic region chr8:1000000-2000000, this genomic region will exactly span the 2 inch width:

```{r philosphy_7, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
bb_pageCreate(width = 3, height = 2, default.units = "inches")

genomicRegion <- bb_params(chrom = "chr8", chromstart = 1000000, chromend = 2000000)

## Plot genes with background color highlighting size
genesPlot <- bb_plotGenes(params = genomicRegion, bg = "light grey",
             x = 0.5, y = 0.5, width = 2, height = 1,
             just = c("left", "top"), default.units = "inches")

## Annotate genome label
bb_annoGenomeLabel(plot = genesPlot, scale = "Mb",
                   x = 0.5, y = 1.5, 
                   just = c("left", "top"), default.units = "inches")
```

When it's time to stack various data types along the same genomic region, users can be confident in accurately aligning containers of the same width:

```{r philosphy_8, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
## Load data
data("bb_hicData")
data("bb_bedpeData")
data("bb_signalData")
## Define genomic region and widths of all plots
params <- bb_params(chrom = "chr21", chromstart = 28000000, chromend = 30300000, 
                    x = unit(0.5, "inches"), width = unit(2, "inches"))

## Create BentoBox page
bb_pageCreate(width = 3, height = 4, default.units = "inches")

## Plot Hi-C data in region
hicPlot <- bb_plotHicSquare(data = bb_hicData,
                            params = params,
                            y = 0.5, height = 2,
                            just = c("left", "top"), default.units = "inches")


## Plot and align DNA loops in region
bedpeLoops <- bb_plotBedpeArches(data = bb_bedpeData, 
                                 params = params, fill = "black", linecolor = "black",
                                 y = 2.6, height = 0.45,
                                 just = c("left", "top"), default.units = "inches")

## Plot and align signal track in region
signalPlot <- bb_plotSignal(data = bb_signalData,
                            params = params,
                            y = 3, height = 0.45,
                            just = c("left", "top"), default.units = "inches")

## Annotate genome label
bb_annoGenomeLabel(plot = signalPlot, 
                   params = params, scale = "Mb",
                   y = 3.6, just = c("left", "top"), default.units = "inches")
```

## Utilizing `grid` graphics `viewports`

`BentoBox` achieves its functionality through `grid` graphics `viewports`, or defined graphical regions. A `BentoBox` page is defined by one large `viewport` with specified dimensions and units. Subsequent plots each have their own `viewport` that are placed on the larger page `viewport`. These `viewports` give users the power to specify the size and placement of each plot's container and clip data to precise axis measurements.
