<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot a gene track for a specified genomic region — plotGenes • plotgardener</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot a gene track for a specified genomic region — plotGenes" />
<meta property="og:description" content="Plot a gene track for a specified genomic region" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plotgardener</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../articles/introduction_to_plotgardener.html">Introduction to plotgardener</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/guides/plotgardener_page.html">The plotgardener Page</a>
    </li>
    <li>
      <a href="../articles/guides/reading_data_for_plotgardener.html">Reading Data for plotgardener</a>
    </li>
    <li>
      <a href="../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../articles/guides/plotgardener_meta_functions.html">plotgardener Meta Functions</a>
    </li>
    <li>
      <a href="../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../articles/guides/incorporating_ggplots.html">Incorporating ggplots</a>
    </li>
    <li>
      <a href="../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/PhanstielLab/plotgardener">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot a gene track for a specified genomic region</h1>
    <small class="dont-index">Source: <a href='https://github.com/PhanstielLab/plotgardener/blob/master/R/plotGenes.R'><code>R/plotGenes.R</code></a></small>
    <div class="hidden name"><code>plotGenes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot a gene track for a specified genomic region</p>
    </div>

    <pre class="usage"><span class='fu'>plotGenes</span><span class='op'>(</span>
    <span class='va'>chrom</span>,
    chromstart <span class='op'>=</span> <span class='cn'>NULL</span>,
    chromend <span class='op'>=</span> <span class='cn'>NULL</span>,
    assembly <span class='op'>=</span> <span class='st'>"hg38"</span>,
    fontsize <span class='op'>=</span> <span class='fl'>8</span>,
    fontcolor <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"#669fd9"</span>, <span class='st'>"#abcc8e"</span><span class='op'>)</span>,
    fill <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"#669fd9"</span>, <span class='st'>"#abcc8e"</span><span class='op'>)</span>,
    geneOrder <span class='op'>=</span> <span class='cn'>NULL</span>,
    geneHighlights <span class='op'>=</span> <span class='cn'>NULL</span>,
    geneBackground <span class='op'>=</span> <span class='st'>"grey"</span>,
    strandLabels <span class='op'>=</span> <span class='cn'>TRUE</span>,
    stroke <span class='op'>=</span> <span class='fl'>0.1</span>,
    bg <span class='op'>=</span> <span class='cn'>NA</span>,
    x <span class='op'>=</span> <span class='cn'>NULL</span>,
    y <span class='op'>=</span> <span class='cn'>NULL</span>,
    width <span class='op'>=</span> <span class='cn'>NULL</span>,
    height <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>0.6</span>, <span class='st'>"inches"</span><span class='op'>)</span>,
    just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>,
    default.units <span class='op'>=</span> <span class='st'>"inches"</span>,
    draw <span class='op'>=</span> <span class='cn'>TRUE</span>,
    params <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>chrom</th>
      <td><p>Chromosome of region to be plotted, as a string.</p></td>
    </tr>
    <tr>
      <th>chromstart</th>
      <td><p>Integer start position on chromosome to be plotted.</p></td>
    </tr>
    <tr>
      <th>chromend</th>
      <td><p>Integer end position on chromosome to be plotted.</p></td>
    </tr>
    <tr>
      <th>assembly</th>
      <td><p>Default genome assembly as a string or a
<a href='assembly.html'>assembly</a> object.
Default value is <code>assembly = "hg38"</code>.</p></td>
    </tr>
    <tr>
      <th>fontsize</th>
      <td><p>A numeric specifying text fontsize in points.
Default value is <code>fontsize = 8</code>.</p></td>
    </tr>
    <tr>
      <th>fontcolor</th>
      <td><p>A character value or vector of length 2 indicating
the fontcolors for the plus strand and minus strand gene labels.
The first value will color the plus strand gene labels and
the second value will color the minus strand gene labels.
Default value is <code>fontcolor = c("#669fd9", "#abcc8e")</code>.</p></td>
    </tr>
    <tr>
      <th>fill</th>
      <td><p>A character value or vector of length 2 indicating the
strand fill colors for the plus strand and minus strand plot elements.
The first value will color the plus strand plot elements and
the second label will color the minus strand plot elements.
Default value is <code>fill = c("#669fd9", "#abcc8e")</code>.</p></td>
    </tr>
    <tr>
      <th>geneOrder</th>
      <td><p>An ordered character vector of gene names to
prioritize when labeling genes.</p></td>
    </tr>
    <tr>
      <th>geneHighlights</th>
      <td><p>A two-column dataframe with a column named "gene"
containing gene names as strings to highlight and a named column "color"
containing corresponding highlight colors.</p></td>
    </tr>
    <tr>
      <th>geneBackground</th>
      <td><p>If <code>geneHighlights</code> is given, a character
value indicating the color for genes that are not highlighted.</p></td>
    </tr>
    <tr>
      <th>strandLabels</th>
      <td><p>A logical value indicating whether to include
+ and - strand labels to the left of the gene track.</p></td>
    </tr>
    <tr>
      <th>stroke</th>
      <td><p>A numeric value indicating the stroke width for gene
body outlines. Default value is <code>stroke = 0.1</code>.</p></td>
    </tr>
    <tr>
      <th>bg</th>
      <td><p>Character value indicating background color.
Default value is <code>bg = NA</code>.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>A numeric or unit object specifying genes plot x-location.</p></td>
    </tr>
    <tr>
      <th>y</th>
      <td><p>A numeric, unit object, or character containing a "b"
combined with a numeric value specifying genes plot y-location.
The character value will
place the genes plot y relative to the bottom of the most recently
plotted plot according to the units of the plotgardener page.</p></td>
    </tr>
    <tr>
      <th>width</th>
      <td><p>A numeric or unit object specifying genes plot width.</p></td>
    </tr>
    <tr>
      <th>height</th>
      <td><p>A numeric or unit object specifying genes plot height.</p></td>
    </tr>
    <tr>
      <th>just</th>
      <td><p>Justification of genes plot relative to its (x, y) location.
If there are two values, the first value specifies horizontal
justification and the second value specifies vertical justification.
Possible string values are: <code>"left"</code>, <code>"right"</code>,
<code>"centre"</code>, <code>"center"</code>, <code>"bottom"</code>, and <code>"top"</code>.
Default value is <code>just = c("left", "top")</code>.</p></td>
    </tr>
    <tr>
      <th>default.units</th>
      <td><p>A string indicating the default units to use if
<code>x</code>, <code>y</code>, <code>width</code>, or <code>height</code> are only given
as numerics. Default value is <code>default.units = "inches"</code>.</p></td>
    </tr>
    <tr>
      <th>draw</th>
      <td><p>A logical value indicating whether graphics output
should be produced. Default value is <code>draw = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>params</th>
      <td><p>An optional <a href='pgParams.html'>pgParams</a> object
containing relevant function parameters.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a <code>genes</code> object containing
relevant genomic region, placement, and <a href='https://rdrr.io/r/grid/grid.grob.html'>grob</a> information.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>A gene track can be placed on a page by providing
plot placement parameters:</p><pre>
<span class='fu'>plotGenes</span><span class='op'>(</span><span class='va'>chrom</span>, chromstart <span class='op'>=</span> <span class='cn'>NULL</span>, chromend <span class='op'>=</span> <span class='cn'>NULL</span>,
            <span class='va'>x</span>, <span class='va'>y</span>, <span class='va'>width</span>, <span class='va'>height</span>, just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>,
            default.units <span class='op'>=</span> <span class='st'>"inches"</span><span class='op'>)</span>
</pre><p>This function can be used to quickly plot an unnannotated gene track
by ignoring plot placement parameters:</p><pre>
<span class='fu'>plotGenes</span><span class='op'>(</span><span class='va'>chrom</span>, chromstart <span class='op'>=</span> <span class='cn'>NULL</span>, chromend <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span>
</pre>

<p>Genomic annotation information is acquired through
<a href='https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html'>TxDb</a> and <a href='https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html'>OrgDb-class</a>
packages, as determined
through the <code>assembly</code> parameter. To avoid overcrowding of gene name
labels, plotted gene labels are by default prioritized according to
citation counts.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='assembly.html'>assembly</a>,
<a href='genomes.html'>genomes</a>, <a href='defaultPackages.html'>defaultPackages</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Load hg19 genomic annotation packages</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'>"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'>"org.Hs.eg.db"</span><span class='op'>)</span>

<span class='co'>## Set genomic coordinates</span>
<span class='va'>paramssmall</span> <span class='op'>&lt;-</span> <span class='fu'><a href='pgParams.html'>pgParams</a></span><span class='op'>(</span>
    chrom <span class='op'>=</span> <span class='st'>"chr8"</span>,
    chromstart <span class='op'>=</span> <span class='fl'>1</span>, chromend <span class='op'>=</span> <span class='fl'>3000000</span>,
    assembly <span class='op'>=</span> <span class='st'>"hg19"</span>, width <span class='op'>=</span> <span class='fl'>7</span>
<span class='op'>)</span>
<span class='va'>paramsbig</span> <span class='op'>&lt;-</span> <span class='fu'><a href='pgParams.html'>pgParams</a></span><span class='op'>(</span>
    chrom <span class='op'>=</span> <span class='st'>"chr8"</span>,
    chromstart <span class='op'>=</span> <span class='fl'>1</span>, chromend <span class='op'>=</span> <span class='fl'>146364022</span>,
    assembly <span class='op'>=</span> <span class='st'>"hg19"</span>, width <span class='op'>=</span> <span class='fl'>7</span>
<span class='op'>)</span>
<span class='co'>## Set colors</span>
<span class='va'>cols</span> <span class='op'>&lt;-</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"#41B6C4"</span>, <span class='st'>"#225EA8"</span><span class='op'>)</span>

<span class='co'>## Create page</span>
<span class='fu'><a href='pageCreate.html'>pageCreate</a></span><span class='op'>(</span>width <span class='op'>=</span> <span class='fl'>7.5</span>, height <span class='op'>=</span> <span class='fl'>3.5</span>, default.units <span class='op'>=</span> <span class='st'>"inches"</span><span class='op'>)</span>
</div><div class='input'>
<span class='co'>## Plot genes big</span>
<span class='va'>genesPlot</span> <span class='op'>&lt;-</span> <span class='fu'>plotGenes</span><span class='op'>(</span>
    params <span class='op'>=</span> <span class='va'>paramsbig</span>, fill <span class='op'>=</span> <span class='va'>cols</span>,
    fontcolor <span class='op'>=</span> <span class='va'>cols</span>,
    x <span class='op'>=</span> <span class='fl'>0.25</span>, y <span class='op'>=</span> <span class='fl'>0.25</span>, height <span class='op'>=</span> <span class='fl'>0.75</span>,
    just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>,
    default.units <span class='op'>=</span> <span class='st'>"inches"</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>genes[genes1]</span></div><div class='input'>
<span class='co'>## Annotate genome label</span>
<span class='fu'><a href='annoGenomeLabel.html'>annoGenomeLabel</a></span><span class='op'>(</span>
    plot <span class='op'>=</span> <span class='va'>genesPlot</span>, x <span class='op'>=</span> <span class='fl'>0.25</span>, y <span class='op'>=</span> <span class='fl'>1.0</span>,
    scale <span class='op'>=</span> <span class='st'>"Mb"</span>, just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: Start label is rounded.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: End label is rounded.</span></div><div class='output co'>#&gt; <span class='message'>genomeLabel[genomeLabel1]</span></div><div class='input'>
<span class='co'>## Plot genes small</span>
<span class='va'>genesPlot</span> <span class='op'>&lt;-</span> <span class='fu'>plotGenes</span><span class='op'>(</span>
    params <span class='op'>=</span> <span class='va'>paramssmall</span>,
    geneHighlights <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
        <span class='st'>"gene"</span> <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"DLGAP2"</span><span class='op'>)</span>,
        <span class='st'>"color"</span> <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"#225EA8"</span><span class='op'>)</span>
    <span class='op'>)</span>,
    geneBackground <span class='op'>=</span> <span class='st'>"grey"</span>,
    x <span class='op'>=</span> <span class='fl'>0.25</span>, y <span class='op'>=</span> <span class='fl'>2.25</span>, height <span class='op'>=</span> <span class='fl'>0.75</span>,
    just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>, default.units <span class='op'>=</span> <span class='st'>"inches"</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>genes[genes2]</span></div><div class='input'>
<span class='co'>## Annotate genome label</span>
<span class='fu'><a href='annoGenomeLabel.html'>annoGenomeLabel</a></span><span class='op'>(</span>
    plot <span class='op'>=</span> <span class='va'>genesPlot</span>, x <span class='op'>=</span> <span class='fl'>0.25</span>, y <span class='op'>=</span> <span class='fl'>3.0</span>, scale <span class='op'>=</span> <span class='st'>"Mb"</span>,
    just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: Start label is rounded.</span></div><div class='output co'>#&gt; <span class='message'>genomeLabel[genomeLabel2]</span></div><div class='img'><img src='plotGenes-1.png' alt='' width='1230' height='1050' /></div><div class='input'>
<span class='co'>## Hide page guides</span>
<span class='fu'><a href='pageGuideHide.html'>pageGuideHide</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='img'><img src='plotGenes-2.png' alt='' width='1230' height='1050' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/nekramer'>Nicole Kramer</a>, <a href='https://github.com/EricSDavis'>Eric S. Davis</a>, <a href='https://github.com/cwenger'>Craig Wenger</a>, <a href='https://github.com/dphansti'>Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


