<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot gene transcripts in a pileup style for a single chromosome — plotTranscripts • plotgardener</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot gene transcripts in a pileup style for a single chromosome — plotTranscripts" />
<meta property="og:description" content="Plot gene transcripts in a pileup style for a single chromosome" />
<meta property="og:image" content="https://phanstiellab.github.io/plotgardener/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plotgardener</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/guides/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../articles/introduction_to_plotgardener.html">Introduction to plotgardener</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/guides/plotgardener_page.html">The plotgardener Page</a>
    </li>
    <li>
      <a href="../articles/guides/reading_data_for_plotgardener.html">Reading Data for plotgardener</a>
    </li>
    <li>
      <a href="../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../articles/guides/plotgardener_meta_functions.html">plotgardener Meta Functions</a>
    </li>
    <li>
      <a href="../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../articles/guides/incorporating_ggplots.html">Incorporating ggplots</a>
    </li>
    <li>
      <a href="../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/PhanstielLab/plotgardener">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot gene transcripts in a pileup style for a single chromosome</h1>
    <small class="dont-index">Source: <a href='https://github.com/PhanstielLab/plotgardener/blob/master/R/plotTranscripts.R'><code>R/plotTranscripts.R</code></a></small>
    <div class="hidden name"><code>plotTranscripts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot gene transcripts in a pileup style for a single chromosome</p>
    </div>

    <pre class="usage"><span class='fu'>plotTranscripts</span><span class='op'>(</span>
    <span class='va'>chrom</span>,
    chromstart <span class='op'>=</span> <span class='cn'>NULL</span>,
    chromend <span class='op'>=</span> <span class='cn'>NULL</span>,
    assembly <span class='op'>=</span> <span class='st'>"hg38"</span>,
    fill <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"#669fd9"</span>, <span class='st'>"#abcc8e"</span><span class='op'>)</span>,
    colorbyStrand <span class='op'>=</span> <span class='cn'>TRUE</span>,
    strandSplit <span class='op'>=</span> <span class='cn'>FALSE</span>,
    boxHeight <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/grid/unit.html'>unit</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='st'>"mm"</span><span class='op'>)</span>,
    spaceWidth <span class='op'>=</span> <span class='fl'>0.02</span>,
    spaceHeight <span class='op'>=</span> <span class='fl'>0.3</span>,
    limitLabel <span class='op'>=</span> <span class='cn'>TRUE</span>,
    fontsize <span class='op'>=</span> <span class='fl'>8</span>,
    labels <span class='op'>=</span> <span class='st'>"transcript"</span>,
    stroke <span class='op'>=</span> <span class='fl'>0.1</span>,
    bg <span class='op'>=</span> <span class='cn'>NA</span>,
    x <span class='op'>=</span> <span class='cn'>NULL</span>,
    y <span class='op'>=</span> <span class='cn'>NULL</span>,
    width <span class='op'>=</span> <span class='cn'>NULL</span>,
    height <span class='op'>=</span> <span class='cn'>NULL</span>,
    just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>,
    default.units <span class='op'>=</span> <span class='st'>"inches"</span>,
    draw <span class='op'>=</span> <span class='cn'>TRUE</span>,
    params <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>chrom</th>
      <td><p>Chromosome of region to be plotted, as a string.</p></td>
    </tr>
    <tr>
      <th>chromstart</th>
      <td><p>Integer start position on chromosome to be plotted.</p></td>
    </tr>
    <tr>
      <th>chromend</th>
      <td><p>Integer end position on chromosome to be plotted.</p></td>
    </tr>
    <tr>
      <th>assembly</th>
      <td><p>Default genome assembly as a string or a
<a href='assembly.html'>assembly</a> object.
Default value is <code>assembly = "hg38"</code>.</p></td>
    </tr>
    <tr>
      <th>fill</th>
      <td><p>Character value(s) as a single value or vector
specifying fill colors of transcripts.
Default value is <code>fill = c("#669fd9", "#abcc8e")</code>.</p></td>
    </tr>
    <tr>
      <th>colorbyStrand</th>
      <td><p>A logical value indicating whether to
color plus and minus strands by the first two colors in
a <code>fill</code> vector, where plus strand transcripts will be
colored by the first <code>fill</code> color and
minus strand transcripts will be colored by the second <code>fill</code>
color. Default value is <code>colorbyStrand = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>strandSplit</th>
      <td><p>A logical value indicating whether plus and
minus-stranded transcripts should be separated, with plus strand
transcripts plotted above the x-axis and minus strand transcripts
plotted below the x-axis. Default value is <code>strandSplit = FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>boxHeight</th>
      <td><p>A numeric or unit object specifying height of transcripts.
Default value is <code>boxHeight = unit(2, "mm")</code>.</p></td>
    </tr>
    <tr>
      <th>spaceWidth</th>
      <td><p>A numeric value specifying the width of minimum spacing
between transcripts, as a fraction of the plot's genomic range.
Default value is <code>spaceWidth = 0.02</code>.</p></td>
    </tr>
    <tr>
      <th>spaceHeight</th>
      <td><p>A numeric value specifying the height of spacing
between transcripts on different rows, as a fraction of <code>boxHeight</code>.
Default value is <code>spaceHeight = 0.3</code>.</p></td>
    </tr>
    <tr>
      <th>limitLabel</th>
      <td><p>A logical value indicating whether to draw a "+"
when not all elements can be plotted in the plotting space. Default 
value is <code>limitLabel = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>fontsize</th>
      <td><p>A numeric specifying text fontsize in points.
Default value is <code>fontsize = 8</code>.</p></td>
    </tr>
    <tr>
      <th>labels</th>
      <td><p>A character value describing the format of
transcript text labels. Default value is <code>labels = "trancript"</code>.
Options are:</p><ul>
<li><p><code>NULL</code>: No labels.</p></li>
<li><p><code>"transcript"</code>: Transcript name labels.</p></li>
<li><p><code>"gene"</code>: Gene name labels.</p></li>
<li><p><code>"both"</code>: Combined transcript and gene name labels with
the format "gene name:transcript name".</p></li>
</ul></td>
    </tr>
    <tr>
      <th>stroke</th>
      <td><p>A numeric value indicating the stroke width for
transcript body outlines. Default value is <code>stroke = 0.1</code>.</p></td>
    </tr>
    <tr>
      <th>bg</th>
      <td><p>Character value indicating background color.
Default value is <code>bg = NA</code>.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>A numeric or unit object specifying transcript plot x-location.</p></td>
    </tr>
    <tr>
      <th>y</th>
      <td><p>A numeric, unit object, or character containing a "b"
combined with a numeric value specifying transcript plot y-location.
The character value will
place the transcript plot y relative to the bottom of the most recently
plotted plot according to the units of the plotgardener page.</p></td>
    </tr>
    <tr>
      <th>width</th>
      <td><p>A numeric or unit object specifying transcript plot width.</p></td>
    </tr>
    <tr>
      <th>height</th>
      <td><p>A numeric or unit object specifying transcript plot height.</p></td>
    </tr>
    <tr>
      <th>just</th>
      <td><p>Justification of transcript plot relative to
its (x, y) location. If there are two values, the first value specifies
horizontal justification and the second value specifies vertical
justification.
Possible string values are: <code>"left"</code>, <code>"right"</code>,
<code>"centre"</code>, <code>"center"</code>, <code>"bottom"</code>, and <code>"top"</code>.
Default value is <code>just = c("left", "top")</code>.</p></td>
    </tr>
    <tr>
      <th>default.units</th>
      <td><p>A string indicating the default units to use if
<code>x</code>, <code>y</code>, <code>width</code>, or <code>height</code> are only given as
numerics. Default value is <code>default.units = "inches"</code>.</p></td>
    </tr>
    <tr>
      <th>draw</th>
      <td><p>A logical value indicating whether graphics output should be
produced. Default value is <code>draw = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>params</th>
      <td><p>An optional <a href='pgParams.html'>pgParams</a> object containing
relevant function parameters.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a <code>transcripts</code> object containing relevant
genomic region, placement, and <a href='https://rdrr.io/r/grid/grid.grob.html'>grob</a> information.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>A transcripts plot can be placed on a plotgardener coordinate page
by providing plot placement parameters:</p><pre>
<span class='fu'>plotTranscripts</span><span class='op'>(</span><span class='va'>chrom</span>, chromstart <span class='op'>=</span> <span class='cn'>NULL</span>, chromend <span class='op'>=</span> <span class='cn'>NULL</span>,
                <span class='va'>x</span>, <span class='va'>y</span>, <span class='va'>width</span>, <span class='va'>height</span>, just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>,
                default.units <span class='op'>=</span> <span class='st'>"inches"</span><span class='op'>)</span>
</pre><p>This function can also be used to quickly plot an unannotated
transcripts plot by ignoring plot placement parameters:</p><pre>
<span class='fu'>plotTranscripts</span><span class='op'>(</span><span class='va'>chrom</span>, chromstart <span class='op'>=</span> <span class='cn'>NULL</span>, chromend <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span>
</pre>

<p>Genomic annotation information is acquired through
<a href='https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html'>TxDb</a> and <a href='https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html'>OrgDb-class</a> packages,
as determined through the <code>assembly</code> parameter.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='assembly.html'>assembly</a>,
<a href='genomes.html'>genomes</a>, <a href='defaultPackages.html'>defaultPackages</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Load hg19 genomic annotation packages</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'>"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'>"org.Hs.eg.db"</span><span class='op'>)</span>

<span class='co'>## Create page</span>
<span class='fu'><a href='pageCreate.html'>pageCreate</a></span><span class='op'>(</span>width <span class='op'>=</span> <span class='fl'>7.5</span>, height <span class='op'>=</span> <span class='fl'>3.5</span>, default.units <span class='op'>=</span> <span class='st'>"inches"</span><span class='op'>)</span>
</div><div class='input'>
<span class='co'>## Plot and place transcripts</span>
<span class='fu'>plotTranscripts</span><span class='op'>(</span>
    chrom <span class='op'>=</span> <span class='st'>"chr8"</span>, chromstart <span class='op'>=</span> <span class='fl'>1000000</span>, chromend <span class='op'>=</span> <span class='fl'>2000000</span>,
    assembly <span class='op'>=</span> <span class='st'>"hg19"</span>, labels <span class='op'>=</span> <span class='st'>"gene"</span>,
    x <span class='op'>=</span> <span class='fl'>0.5</span>, y <span class='op'>=</span> <span class='fl'>0.5</span>, width <span class='op'>=</span> <span class='fl'>6.5</span>, height <span class='op'>=</span> <span class='fl'>2.5</span>,
    just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>, default.units <span class='op'>=</span> <span class='st'>"inches"</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>transcripts[transcripts1]</span></div><div class='input'>
<span class='co'>## Plot genome label</span>
<span class='fu'><a href='plotGenomeLabel.html'>plotGenomeLabel</a></span><span class='op'>(</span>
    chrom <span class='op'>=</span> <span class='st'>"chr8"</span>, chromstart <span class='op'>=</span> <span class='fl'>1000000</span>, chromend <span class='op'>=</span> <span class='fl'>2000000</span>,
    assembly <span class='op'>=</span> <span class='st'>"hg19"</span>,
    x <span class='op'>=</span> <span class='fl'>0.5</span>, y <span class='op'>=</span> <span class='fl'>3.03</span>, length <span class='op'>=</span> <span class='fl'>6.5</span>, default.units <span class='op'>=</span> <span class='st'>"inches"</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>genomeLabel[genomeLabel1]</span></div><div class='input'>
<span class='co'>## Plot a legend</span>
<span class='fu'><a href='plotLegend.html'>plotLegend</a></span><span class='op'>(</span>
    legend <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"+ strand"</span>, <span class='st'>"- strand"</span><span class='op'>)</span>,
    fill <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"#669fd9"</span>, <span class='st'>"#abcc8e"</span><span class='op'>)</span>, border <span class='op'>=</span> <span class='cn'>FALSE</span>,
    x <span class='op'>=</span> <span class='fl'>0.5</span>, y <span class='op'>=</span> <span class='fl'>1</span>, width <span class='op'>=</span> <span class='fl'>1</span>, height <span class='op'>=</span> <span class='fl'>0.5</span>,
    just <span class='op'>=</span> <span class='fu'><a href='c.html'>c</a></span><span class='op'>(</span><span class='st'>"left"</span>, <span class='st'>"top"</span><span class='op'>)</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>legend[legend1]</span></div><div class='img'><img src='plotTranscripts-1.png' alt='' width='1230' height='1050' /></div><div class='input'>
<span class='co'>## Hide page guides</span>
<span class='fu'><a href='pageGuideHide.html'>pageGuideHide</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='img'><img src='plotTranscripts-2.png' alt='' width='1230' height='1050' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/nekramer'>Nicole Kramer</a>, <a href='https://github.com/EricSDavis'>Eric S. Davis</a>, <a href='https://github.com/cwenger'>Craig Wenger</a>, <a href='https://github.com/dphansti'>Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


