<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot genomic coordinates along the x or y-axis of a plotgardener plot — plotGenomeLabel • plotgardener</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Plot genomic coordinates along the x or y-axis of a plotgardener plot — plotGenomeLabel"><meta property="og:description" content="Plot genomic coordinates along the x or y-axis of a plotgardener plot"><meta property="og:image" content="https://phanstiellab.github.io/plotgardener/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plotgardener</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/guides/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../articles/introduction_to_plotgardener.html">Introduction to plotgardener</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/guides/plotgardener_page.html">The plotgardener Page</a>
    </li>
    <li>
      <a href="../articles/guides/reading_data_for_plotgardener.html">Reading Data for plotgardener</a>
    </li>
    <li>
      <a href="../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../articles/guides/plotgardener_meta_functions.html">plotgardener Meta Functions</a>
    </li>
    <li>
      <a href="../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../articles/guides/incorporating_ggplots.html">Incorporating ggplots and other grid-based Bioconductor visualizations</a>
    </li>
    <li>
      <a href="../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/PhanstielLab/plotgardener" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot genomic coordinates along the x or y-axis of a plotgardener plot</h1>
    <small class="dont-index">Source: <a href="https://github.com/PhanstielLab/plotgardener/blob/HEAD/R/plotGenomeLabel.R" class="external-link"><code>R/plotGenomeLabel.R</code></a></small>
    <div class="hidden name"><code>plotGenomeLabel.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot genomic coordinates along the x or y-axis of a plotgardener plot</p>
    </div>

    <div id="ref-usage">Usage,<div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plotGenomeLabel</span><span class="op">(</span>
    <span class="va">chrom</span>,
    chromstart <span class="op">=</span> <span class="cn">NULL</span>,
    chromend <span class="op">=</span> <span class="cn">NULL</span>,
    assembly <span class="op">=</span> <span class="st">"hg38"</span>,
    fontsize <span class="op">=</span> <span class="fl">10</span>,
    fontcolor <span class="op">=</span> <span class="st">"black"</span>,
    linecolor <span class="op">=</span> <span class="st">"black"</span>,
    margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"mm"</span><span class="op">)</span>,
    scale <span class="op">=</span> <span class="st">"bp"</span>,
    commas <span class="op">=</span> <span class="cn">TRUE</span>,
    sequence <span class="op">=</span> <span class="cn">TRUE</span>,
    boxWidth <span class="op">=</span> <span class="fl">0.5</span>,
    axis <span class="op">=</span> <span class="st">"x"</span>,
    at <span class="op">=</span> <span class="cn">NULL</span>,
    tcl <span class="op">=</span> <span class="fl">0.5</span>,
    <span class="va">x</span>,
    <span class="va">y</span>,
    <span class="va">length</span>,
    just <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,
    default.units <span class="op">=</span> <span class="st">"inches"</span>,
    params <span class="op">=</span> <span class="cn">NULL</span>,
    <span class="va">...</span>
<span class="op">)</span></code></pre></div></div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>chrom</dt>
<dd><p>Chromosome of genome label, as a string,
or a character vector of chromosomes for a whole genome Manhattan plot.</p></dd>
<dt>chromstart</dt>
<dd><p>Integer start of genome label.</p></dd>
<dt>chromend</dt>
<dd><p>Integer end of genome label.</p></dd>
<dt>assembly</dt>
<dd><p>Default genome assembly as a string or a
<a href="assembly.html">assembly</a> object.</p></dd>
<dt>fontsize</dt>
<dd><p>A numeric specifying text fontsize in points.
Default value is <code>fontsize = 10</code>.</p></dd>
<dt>fontcolor</dt>
<dd><p>A character value indicating the color for text.
Default value is <code>fontcolor = "black"</code>.</p></dd>
<dt>linecolor</dt>
<dd><p>A character value indicating the color of
the genome label axis. Default value is <code>linecolor = "black"</code>.</p></dd>
<dt>margin</dt>
<dd><p>A numeric or unit vector specifying space between axis
and coordinate labels. Default value is <code>margin = unit(1, "mm")</code>,</p></dd>
<dt>scale</dt>
<dd><p>A character value indicating the scale of the coordinates
along the genome label. Default value is <code>scale = "bp"</code>. Options are:</p><ul><li><p><code>"bp"</code>: base pairs.</p></li>
<li><p><code>"Kb"</code>: kilobase pairs. 1 kilobase pair is equal to
1000 base pairs.</p></li>
<li><p><code>"Mb"</code>: megabase pairs. 1 megabase pair is equal to
1000000 base pairs.</p></li>
</ul></dd>
<dt>commas</dt>
<dd><p>A logical value indicating whether to include commas in
start and stop labels. Default value is <code>commas = TRUE</code>.</p></dd>
<dt>sequence</dt>
<dd><p>A logical value indicating whether to include sequence
information above the label of an x-axis (only at appropriate resolutions).</p></dd>
<dt>boxWidth</dt>
<dd><p>A numeric value indicating the width of the boxes
representing sequence information at appropriate resolutions.
Default value is <code>boxWidth = 0.5</code>.</p></dd>
<dt>axis</dt>
<dd><p>A character value indicating along which axis to
add genome label. Sequence information will not be displayed along a y-axis.
Default value is <code>axis = "x"</code>.
Options are:</p><ul><li><p><code>"x"</code>: Genome label will be plotted along the x-axis.</p></li>
<li><p><code>"y"</code>: Genome label will be plotted along the y-axis.
This is typically used for a square Hi-C plot made with
<code>plotHicSquare</code>.</p></li>
</ul></dd>
<dt>at</dt>
<dd><p>A numeric vector of x-value locations for tick marks.</p></dd>
<dt>tcl</dt>
<dd><p>A numeric specifying the length of tickmarks as a
fraction of text height. Default value is <code>tcl = 0.5</code>.</p></dd>
<dt>x</dt>
<dd><p>A numeric or unit object specifying genome label x-location.</p></dd>
<dt>y</dt>
<dd><p>A numeric, unit object, or character containing a "b"
combined with a numeric value specifying genome label y-location.
The character value will
place the genome label y relative to the bottom of the most recently
plotted plot according to the units of the plotgardener page.</p></dd>
<dt>length</dt>
<dd><p>A numeric or unit object specifying length of
genome label axis.</p></dd>
<dt>just</dt>
<dd><p>Justification of genome label relative to its (x, y)
location. If there are two values, the first value specifies horizontal
justification and the second value specifies vertical justification.
Possible string values are: <code>"left"</code>, <code>"right"</code>,
<code>"centre"</code>, <code>"center"</code>, <code>"bottom"</code>,
and <code>"top"</code>. Default value is <code>just = c("left", "top")</code>.</p></dd>
<dt>default.units</dt>
<dd><p>A string indicating the default units to
use if <code>x</code>, <code>y</code>, or <code>length</code> are only given as numerics.
Default value is <code>default.units = "inches"</code>.</p></dd>
<dt>params</dt>
<dd><p>An optional <a href="pgParams.html">pgParams</a> object
containing relevant function parameters.</p></dd>
<dt>...</dt>
<dd><p>Additional grid graphical parameters or digit specifications.
See <a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a> and <a href="https://rdrr.io/r/base/formatc.html" class="external-link">formatC</a>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Returns a <code>genomeLabel</code> object containing
relevant genomic region, placement, and <a href="https://rdrr.io/r/grid/grid.grob.html" class="external-link">grob</a> information.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co">## Load hg19 genomic annotation packages</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class="op">)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BSgenome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biostrings</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: XVector</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biostrings’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:grid’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pattern</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     strsplit</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: rtracklayer</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Create page</span></span>
<span class="r-in"><span class="fu"><a href="pageCreate.html">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">5</span>, height <span class="op">=</span> <span class="fl">3</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Plot and place gene track on page</span></span>
<span class="r-in"><span class="va">genesPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="plotGenes.html">plotGenes</a></span><span class="op">(</span></span>
<span class="r-in">    chrom <span class="op">=</span> <span class="st">"chr8"</span>,</span>
<span class="r-in">    chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">2000000</span>,</span>
<span class="r-in">    assembly <span class="op">=</span> <span class="st">"hg19"</span>, fill <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"grey"</span><span class="op">)</span>,</span>
<span class="r-in">    fontcolor <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"grey"</span><span class="op">)</span>,</span>
<span class="r-in">    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">4</span>, height <span class="op">=</span> <span class="fl">1</span>,</span>
<span class="r-in">    just <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,</span>
<span class="r-in">    default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> genes[genes1]</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Plot x-axis genome labels at different scales</span></span>
<span class="r-in"><span class="fu">plotGenomeLabel</span><span class="op">(</span></span>
<span class="r-in">    chrom <span class="op">=</span> <span class="st">"chr8"</span>,</span>
<span class="r-in">    chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">2000000</span>,</span>
<span class="r-in">    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span class="r-in">    scale <span class="op">=</span> <span class="st">"Mb"</span>,</span>
<span class="r-in">    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1.25</span>, length <span class="op">=</span> <span class="fl">4</span>, just <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,</span>
<span class="r-in">    default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> genomeLabel[genomeLabel1]</span>
<span class="r-in"><span class="fu">plotGenomeLabel</span><span class="op">(</span></span>
<span class="r-in">    chrom <span class="op">=</span> <span class="st">"chr8"</span>,</span>
<span class="r-in">    chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">2000000</span>,</span>
<span class="r-in">    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span class="r-in">    scale <span class="op">=</span> <span class="st">"Kb"</span>,</span>
<span class="r-in">    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1.5</span>, length <span class="op">=</span> <span class="fl">4</span>, just <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,</span>
<span class="r-in">    default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> genomeLabel[genomeLabel2]</span>
<span class="r-in"><span class="fu">plotGenomeLabel</span><span class="op">(</span></span>
<span class="r-in">    chrom <span class="op">=</span> <span class="st">"chr8"</span>,</span>
<span class="r-in">    chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">2000000</span>,</span>
<span class="r-in">    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span class="r-in">    scale <span class="op">=</span> <span class="st">"bp"</span>,</span>
<span class="r-in">    x <span class="op">=</span> <span class="fl">0.5</span>, y <span class="op">=</span> <span class="fl">1.75</span>, length <span class="op">=</span> <span class="fl">4</span>, just <span class="op">=</span> <span class="fu"><a href="c.html">c</a></span><span class="op">(</span><span class="st">"left"</span>, <span class="st">"top"</span><span class="op">)</span>,</span>
<span class="r-in">    default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> genomeLabel[genomeLabel3]</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Plot a different genomic label region, zooming in enough</span></span>
<span class="r-in"><span class="co">## to see base pairs</span></span>
<span class="r-in"><span class="fu">plotGenomeLabel</span><span class="op">(</span></span>
<span class="r-in">    chrom <span class="op">=</span> <span class="st">"chr8"</span>,</span>
<span class="r-in">    chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">1000050</span>,</span>
<span class="r-in">    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span class="r-in">    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">2.2</span>, length <span class="op">=</span> <span class="fl">4.5</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> genomeLabel[genomeLabel4]</span>
<span class="r-in"><span class="fu">plotGenomeLabel</span><span class="op">(</span></span>
<span class="r-in">    chrom <span class="op">=</span> <span class="st">"chr8"</span>,</span>
<span class="r-in">    chromstart <span class="op">=</span> <span class="fl">1000000</span>, chromend <span class="op">=</span> <span class="fl">1000020</span>,</span>
<span class="r-in">    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span class="r-in">    x <span class="op">=</span> <span class="fl">0</span>, y <span class="op">=</span> <span class="fl">2.6</span>, length <span class="op">=</span> <span class="fl">5</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> genomeLabel[genomeLabel6]</span>
<span class="r-plt img"><img src="plotGenomeLabel-1.png" alt="" width="1230" height="1050"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Hide page guides</span></span>
<span class="r-in"><span class="fu"><a href="pageGuideHide.html">pageGuideHide</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-plt img"><img src="plotGenomeLabel-2.png" alt="" width="1230" height="1050"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/nekramer" class="external-link">Nicole Kramer</a>, <a href="https://github.com/EricSDavis" class="external-link">Eric S. Davis</a>, <a href="https://github.com/cwenger" class="external-link">Craig Wenger</a>, <a href="https://github.com/dphansti" class="external-link">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.0.</p>
</div>

      </footer></div>

  


  

  </body></html>

