<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bioconductor Integration • BentoBox</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Bioconductor Integration">
<meta property="og:description" content="BentoBox">
<meta property="og:image" content="https://phanstiellab.github.io/bentobox/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BentoBox</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../articles/introduction_to_bentobox.html">Introduction to BentoBox</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/bentobox_page.html">The BentoBox Page</a>
    </li>
    <li>
      <a href="../articles/reading_data_for_bentobox.html">Reading Data for BentoBox</a>
    </li>
    <li>
      <a href="../articles/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../articles/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../articles/bentobox_meta_functions.html">BentoBox Meta Functions</a>
    </li>
    <li>
      <a href="../articles/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../articles/incorporating_ggplots.html">Incorporating ggplots</a>
    </li>
    <li>
      <a href="../articles/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../articles/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhanstielLab/BentoBox">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="bioconductor_integration_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Bioconductor Integration</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PhanstielLab/BentoBox/blob/master/vignettes/bioconductor_integration.Rmd"><code>vignettes/bioconductor_integration.Rmd</code></a></small>
      <div class="hidden name"><code>bioconductor_integration.Rmd</code></div>

    </div>

    
    
<p><code>BentoBox</code> is designed to be flexibly compatible with typical Bioconductor classes and libraries of genomic data to easily integrate genomic data analysis and visualization. In addition to handling various genomic file types and <code>R</code> objects, many <code>BentoBox</code> functions can also handle <code>GRanges</code> or <code>GInteractions</code> objects as input data. Furthermore, <code>BentoBox</code> does not hard-code any genomic assemblies and can utilize <code>TxDb</code>, <code>OrgDb</code>, and <code>BSgenome</code> objects for various genomic annotations, including gene and transcript structures and names, chromosome sizes, and nucleotide sequences. For standard genomic assemblies (i.e. hg19, hg38, mm10), <code>BentoBox</code> uses a set of default packages that can be displayed by calling <code><a href="../reference/bbDefaultPackages.html">bbDefaultPackages()</a></code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bbDefaultPackages.html">bbDefaultPackages</a></span><span class="op">(</span><span class="st">"hg38"</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    1 obs. of  6 variables:</span>
<span class="co">#&gt;  $ Genome        : chr "hg38"</span>
<span class="co">#&gt;  $ TxDb          : chr "TxDb.Hsapiens.UCSC.hg38.knownGene"</span>
<span class="co">#&gt;  $ OrgDb         : chr "org.Hs.eg.db"</span>
<span class="co">#&gt;  $ gene.id.column: chr "ENTREZID"</span>
<span class="co">#&gt;  $ display.column: chr "SYMBOL"</span>
<span class="co">#&gt;  $ BSgenome      : chr "BSgenome.Hsapiens.UCSC.hg38"</span>
<span class="fu"><a href="../reference/bbDefaultPackages.html">bbDefaultPackages</a></span><span class="op">(</span><span class="st">"hg19"</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    1 obs. of  6 variables:</span>
<span class="co">#&gt;  $ Genome        : chr "hg19"</span>
<span class="co">#&gt;  $ TxDb          : chr "TxDb.Hsapiens.UCSC.hg19.knownGene"</span>
<span class="co">#&gt;  $ OrgDb         : chr "org.Hs.eg.db"</span>
<span class="co">#&gt;  $ gene.id.column: chr "ENTREZID"</span>
<span class="co">#&gt;  $ display.column: chr "SYMBOL"</span>
<span class="co">#&gt;  $ BSgenome      : chr "BSgenome.Hsapiens.UCSC.hg19"</span>
<span class="fu"><a href="../reference/bbDefaultPackages.html">bbDefaultPackages</a></span><span class="op">(</span><span class="st">"mm10"</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    1 obs. of  6 variables:</span>
<span class="co">#&gt;  $ Genome        : chr "mm10"</span>
<span class="co">#&gt;  $ TxDb          : chr "TxDb.Mmusculus.UCSC.mm10.knownGene"</span>
<span class="co">#&gt;  $ OrgDb         : chr "org.Mm.eg.db"</span>
<span class="co">#&gt;  $ gene.id.column: chr "ENTREZID"</span>
<span class="co">#&gt;  $ display.column: chr "SYMBOL"</span>
<span class="co">#&gt;  $ BSgenome      : chr "BSgenome.Mmusculus.UCSC.mm10"</span></code></pre></div>
<p>To see which assemblies have defaults within <code>BentoBox</code>, call <code><a href="../reference/bbGenomes.html">bbGenomes()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bbGenomes.html">bbGenomes</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; bosTau8</span>
<span class="co">#&gt; bosTau9</span>
<span class="co">#&gt; canFam3</span>
<span class="co">#&gt; ce6</span>
<span class="co">#&gt; ce11</span>
<span class="co">#&gt; danRer10</span>
<span class="co">#&gt; danRer11</span>
<span class="co">#&gt; dm3</span>
<span class="co">#&gt; dm6</span>
<span class="co">#&gt; galGal4</span>
<span class="co">#&gt; galGal5</span>
<span class="co">#&gt; galGal6</span>
<span class="co">#&gt; hg18</span>
<span class="co">#&gt; hg19</span>
<span class="co">#&gt; hg38</span>
<span class="co">#&gt; mm9</span>
<span class="co">#&gt; mm10</span>
<span class="co">#&gt; rheMac3</span>
<span class="co">#&gt; rheMac8</span>
<span class="co">#&gt; rheMac10</span>
<span class="co">#&gt; panTro5</span>
<span class="co">#&gt; panTro6</span>
<span class="co">#&gt; rn4</span>
<span class="co">#&gt; rn5</span>
<span class="co">#&gt; rn6</span>
<span class="co">#&gt; sacCer2</span>
<span class="co">#&gt; sacCer3</span>
<span class="co">#&gt; susScr3</span>
<span class="co">#&gt; susScr11</span></code></pre></div>
<p><code>BentoBox</code> functions default to an “hg38” assembly, but can be customized with any of the other genomic assemblies included or a <code>bbAssembly</code> object. To create custom genomic assemblies and combinations of <code>TxDb</code>, <code>orgDb</code>, and <code>BSgenome</code> packages for use in <code>BentoBox</code> functions, we can use the <code><a href="../reference/bbAssembly.html">bbAssembly()</a></code> constructor. For example, we can create our own <code>TxDb</code> from the current human Ensembl release:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures">GenomicFeatures</a></span><span class="op">)</span>
<span class="va">TxDb.Hsapiens.Ensembl.GrCh38.103</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromEnsembl.html">makeTxDbFromEnsembl</a></span><span class="op">(</span>
    organism <span class="op">=</span>
        <span class="st">"Homo sapiens"</span>
<span class="op">)</span></code></pre></div>
<p>We can now create a new <code>bbAssembly</code> with this <code>TxDb</code> and combinations of other Bioconductor packages. The <code>Genome</code> parameter can is a string to name or describe this assembly. Since the <code>TxDb</code> is now from ENSEMBL, we will change the <code>gene.id</code> field to <code>"ENSEMBL"</code> to map gene IDs and symbols between our <code>TxDb</code> and <code>orgDb</code> objects. Most gene ID types can be found by calling <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">AnnotationDbi::keytypes()</a></code> on an <code>orgDb</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Ensembl38</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bbAssembly.html">bbAssembly</a></span><span class="op">(</span>
    Genome <span class="op">=</span> <span class="st">"Ensembl.GRCh38.103"</span>,
    TxDb <span class="op">=</span> <span class="va">TxDb.Hsapiens.Ensembl.GrCh38.103</span>,
    OrgDb <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>,
    BSgenome <span class="op">=</span> <span class="st">"BSgenome.Hsapiens.NCBI.GRCh38"</span>,
    gene.id <span class="op">=</span> <span class="st">"ENSEMBL"</span>, display.column <span class="op">=</span> <span class="st">"SYMBOL"</span>
<span class="op">)</span></code></pre></div>
<p>This <code>bbAssembly</code> object can now be easily passed into <code>BentoBox</code> functions through the <code>assembly</code> parameter.</p>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] grid      stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] BentoBox_0.99.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] MatrixGenerics_1.4.0        Biobase_2.52.0             </span>
<span class="co">#&gt;  [3] sass_0.4.0                  jsonlite_1.7.2             </span>
<span class="co">#&gt;  [5] bslib_0.2.5.1               assertthat_0.2.1           </span>
<span class="co">#&gt;  [7] BiocManager_1.30.16         rvcheck_0.1.8              </span>
<span class="co">#&gt;  [9] stats4_4.1.0                GenomeInfoDbData_1.2.6     </span>
<span class="co">#&gt; [11] Rsamtools_2.8.0             yaml_2.2.1                 </span>
<span class="co">#&gt; [13] pillar_1.6.2                lattice_0.20-44            </span>
<span class="co">#&gt; [15] glue_1.4.2                  digest_0.6.27              </span>
<span class="co">#&gt; [17] GenomicRanges_1.44.0        RColorBrewer_1.1-2         </span>
<span class="co">#&gt; [19] XVector_0.32.0              colorspace_2.0-2           </span>
<span class="co">#&gt; [21] htmltools_0.5.1.1           Matrix_1.3-4               </span>
<span class="co">#&gt; [23] strawr_0.0.8                XML_3.99-0.6               </span>
<span class="co">#&gt; [25] pkgconfig_2.0.3             zlibbioc_1.38.0            </span>
<span class="co">#&gt; [27] purrr_0.3.4                 scales_1.1.1               </span>
<span class="co">#&gt; [29] ggplotify_0.0.7             BiocParallel_1.26.1        </span>
<span class="co">#&gt; [31] tibble_3.1.3                generics_0.1.0             </span>
<span class="co">#&gt; [33] IRanges_2.26.0              ggplot2_3.3.5              </span>
<span class="co">#&gt; [35] ellipsis_0.3.2              cachem_1.0.5               </span>
<span class="co">#&gt; [37] SummarizedExperiment_1.22.0 BiocGenerics_0.38.0        </span>
<span class="co">#&gt; [39] magrittr_2.0.1              crayon_1.4.1               </span>
<span class="co">#&gt; [41] memoise_2.0.0               evaluate_0.14              </span>
<span class="co">#&gt; [43] fs_1.5.0                    fansi_0.5.0                </span>
<span class="co">#&gt; [45] textshaping_0.3.5           tools_4.1.0                </span>
<span class="co">#&gt; [47] data.table_1.14.0           BiocIO_1.2.0               </span>
<span class="co">#&gt; [49] lifecycle_1.0.0             matrixStats_0.60.0         </span>
<span class="co">#&gt; [51] stringr_1.4.0               plyranges_1.12.1           </span>
<span class="co">#&gt; [53] S4Vectors_0.30.0            munsell_0.5.0              </span>
<span class="co">#&gt; [55] DelayedArray_0.18.0         Biostrings_2.60.1          </span>
<span class="co">#&gt; [57] compiler_4.1.0              pkgdown_1.6.1              </span>
<span class="co">#&gt; [59] jquerylib_0.1.4             GenomeInfoDb_1.28.1        </span>
<span class="co">#&gt; [61] gridGraphics_0.5-1          systemfonts_1.0.2          </span>
<span class="co">#&gt; [63] rlang_0.4.11                RCurl_1.98-1.3             </span>
<span class="co">#&gt; [65] rjson_0.2.20                bitops_1.0-7               </span>
<span class="co">#&gt; [67] rmarkdown_2.9               restfulr_0.0.13            </span>
<span class="co">#&gt; [69] gtable_0.3.0                DBI_1.1.1                  </span>
<span class="co">#&gt; [71] curl_4.3.2                  R6_2.5.0                   </span>
<span class="co">#&gt; [73] GenomicAlignments_1.28.0    knitr_1.33                 </span>
<span class="co">#&gt; [75] dplyr_1.0.7                 rtracklayer_1.52.0         </span>
<span class="co">#&gt; [77] fastmap_1.1.0               utf8_1.2.2                 </span>
<span class="co">#&gt; [79] rprojroot_2.0.2             ragg_1.1.3                 </span>
<span class="co">#&gt; [81] desc_1.3.0                  stringi_1.7.3              </span>
<span class="co">#&gt; [83] parallel_4.1.0              Rcpp_1.0.7                 </span>
<span class="co">#&gt; [85] vctrs_0.3.8                 tidyselect_1.1.1           </span>
<span class="co">#&gt; [87] xfun_0.24</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/nekramer">Nicole Kramer</a>, <a href="https://github.com/EricSDavis">Eric S. Davis</a>, <a href="https://github.com/cwenger">Craig Wenger</a>, <a href="https://github.com/dphansti">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
