<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to plotgardener • plotgardener</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction to plotgardener">
<meta property="og:description" content="plotgardener">
<meta property="og:image" content="https://phanstiellab.github.io/plotgardener/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plotgardener</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/guides/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../articles/introduction_to_plotgardener.html">Introduction to plotgardener</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/guides/plotgardener_page.html">The plotgardener Page</a>
    </li>
    <li>
      <a href="../articles/guides/reading_data_for_plotgardener.html">Reading Data for plotgardener</a>
    </li>
    <li>
      <a href="../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../articles/guides/plotgardener_meta_functions.html">plotgardener Meta Functions</a>
    </li>
    <li>
      <a href="../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../articles/guides/incorporating_ggplots.html">Incorporating ggplots and other grid-based Bioconductor visualizations</a>
    </li>
    <li>
      <a href="../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhanstielLab/plotgardener" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to plotgardener</h1>
                        <h4 data-toc-skip class="author">Nicole Kramer,
Eric S. Davis, Craig Wenger, Sarah Parker, Erika Deoudes, Michael Love,
Douglas H. Phanstiel</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/PhanstielLab/plotgardener/blob/HEAD/vignettes/introduction_to_plotgardener.Rmd" class="external-link"><code>vignettes/introduction_to_plotgardener.Rmd</code></a></small>
      <div class="hidden name"><code>introduction_to_plotgardener.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<img src="../reference/figures/pg-hex-text.png" align="right" width="140px" style="padding-left:20px; background-color:white; border-color: transparent"><p><code>plotgardener</code> is a coordinate-based, genomic
visualization package for R. Using <code>grid</code> graphics,
<code>plotgardener</code> empowers users to programmatically and
flexibly generate multi-panel figures. Tailored for genomics for a
variety of genomic assemblies, <code>plotgardener</code> allows users to
visualize large, complex genomic datasets while providing exquisite
control over the arrangement of plots.</p>
<p><code>plotgardener</code> functions can be grouped into the following
categories:</p>
<ul>
<li><strong>Page layout functions:</strong></li>
</ul>
<p>Functions for creating <code>plotgardener</code> page layouts,
drawing, showing, and hiding guides, as well as placing plots on the
page. See <a href="guides/plotgardener_page.html">The plotgardener
Page</a></p>
<ul>
<li><strong>Reading functions:</strong></li>
</ul>
<p>Functions for quickly reading in large biological datasets. See <a href="guides/reading_data_for_plotgardener.html">Reading Data for
plotgardener</a></p>
<ul>
<li><strong>Plotting functions:</strong></li>
</ul>
<p>Contains genomic plotting functions, functions for placing
<code>ggplots</code> and <code>base</code> plots, as well as functions
for drawing simple shapes. See <a href="guides/plotting_multiomic_data.html">Plotting Multi-omic
Data</a></p>
<ul>
<li><strong>Annotation functions:</strong></li>
</ul>
<p>Enables users to add annotations to their plots, such as legends,
axes, and scales. See <a href="guides/annotations.html">Plot
Annotations</a></p>
<ul>
<li><strong>Meta functions:</strong></li>
</ul>
<p>Functions that display <code>plotgardener</code> properties or
operate on other <code>plotgardener</code> functions, or constructors
for <code>plotgardener</code> objects. See <a href="guides/plotgardener_meta_functions.html">plotgardener Meta
Functions</a></p>
<p>This vignette provides a quick start guide for utilizing
<code>plotgardener</code>. For in-depth demonstrations of
<code>plotgardener</code>’s key features, see the additional articles.
For detailed usage of each function, see the function-specific reference
examples with <code>?function()</code>
(e.g. <code>?plotPairs()</code>).</p>
<p>All the data included in this vignette can be found in the
supplementary package <code>plotgardenerData</code>.</p>
</div>
<div class="section level2">
<h2 id="quick-plotting">Quick plotting<a class="anchor" aria-label="anchor" href="#quick-plotting"></a>
</h2>
<p><code>plotgardener</code> plotting functions contain 4 types of
arguments:</p>
<ol style="list-style-type: decimal">
<li><p>Data reading argument (<code>data</code>)</p></li>
<li><p>Genomic locus arguments (<code>chrom</code>,
<code>chromstart</code>, <code>chromend</code>,
<code>assembly</code>)</p></li>
<li><p>Placement arguments (<code>x</code>, <code>y</code>,
<code>width</code>, <code>height</code>, <code>just</code>,
<code>default.units</code>, …) that define where each plot resides on a
<code>page</code></p></li>
<li><p>Attribute arguments that affect the data being plotted or the
style of the plot (<code>norm</code>, <code>fill</code>,
<code>fontcolor</code>, …) that vary between functions</p></li>
</ol>
<p>The quickest way to plot data is to omit the placement arguments.
This will generate a <code>plotgardener</code> plot that fills up the
entire graphics window and cannot be annotated. <strong>These plots are
only meant to be used for quick</strong> <strong>genomic data inspection
and not as final <code>plotgardener</code> plots.</strong> The only
arguments that are required are the data arguments and locus arguments.
The examples below show how to quickly plot different types of genomic
data with plot defaults and included data types. To use your own data,
replace the <code>data</code> argument with either a path to the file or
an R object as described above.</p>
<div class="section level3">
<h3 id="hi-c-matrices">Hi-C matrices<a class="anchor" aria-label="anchor" href="#hi-c-matrices"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load plotgardener</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener">plotgardener</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load example Hi-C data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/plotgardenerData" class="external-link">plotgardenerData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Quick plot Hi-C data</span></span>
<span><span class="fu"><a href="../reference/plotHicSquare.html">plotHicSquare</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,</span>
<span>    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,</span>
<span>    assembly <span class="op">=</span> <span class="st">"hg19"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/hic_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="signal-tracks">Signal tracks<a class="anchor" aria-label="anchor" href="#signal-tracks"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load plotgardener</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener">plotgardener</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load example signal data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/plotgardenerData" class="external-link">plotgardenerData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_H3K27ac_signal"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Quick plot signal data</span></span>
<span><span class="fu"><a href="../reference/plotSignal.html">plotSignal</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">IMR90_ChIP_H3K27ac_signal</span>,</span>
<span>    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,</span>
<span>    assembly <span class="op">=</span> <span class="st">"hg19"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/signal_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="gene-tracks">Gene tracks<a class="anchor" aria-label="anchor" href="#gene-tracks"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load plotgardener</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener">plotgardener</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load hg19 genomic annotation packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Quick plot genes</span></span>
<span><span class="fu"><a href="../reference/plotGenes.html">plotGenes</a></span><span class="op">(</span></span>
<span>    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/gene_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="gwas-manhattan-plots">GWAS Manhattan plots<a class="anchor" aria-label="anchor" href="#gwas-manhattan-plots"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load plotgardener</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/plotgardener">plotgardener</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load hg19 genomic annotation packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load example GWAS data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/plotgardenerData" class="external-link">plotgardenerData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"hg19_insulin_GWAS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Quick plot GWAS data</span></span>
<span><span class="fu"><a href="../reference/plotManhattan.html">plotManhattan</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">hg19_insulin_GWAS</span>, </span>
<span>    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>    fill <span class="op">=</span> <span class="fu"><a href="../reference/c.html">c</a></span><span class="op">(</span><span class="st">"steel blue"</span>, <span class="st">"grey"</span><span class="op">)</span>,</span>
<span>    ymax <span class="op">=</span> <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">0.20</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/gwas_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="plotting-and-annotating-on-the-plotgardener-page">Plotting and annotating on the <code>plotgardener</code> page<a class="anchor" aria-label="anchor" href="#plotting-and-annotating-on-the-plotgardener-page"></a>
</h2>
<p>To build complex, multi-panel <code>plotgardener</code> figures with
annotations, we must:</p>
<ol style="list-style-type: decimal">
<li>Create a <code>plotgardener</code> coordinate page with
<code><a href="../reference/pageCreate.html">pageCreate()</a></code>.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pageCreate.html">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/quickpage-1.png" width="600" style="display: block; margin: auto;"></p>
<ol start="2" style="list-style-type: decimal">
<li>Provide values for the placement arguments (<code>x</code>,
<code>y</code>, <code>width</code>, <code>height</code>,
<code>just</code>, <code>default.units</code>) in plotting functions and
save the output of the plotting function.</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span></span>
<span><span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHicSquare.html">plotHicSquare</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,</span>
<span>    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,</span>
<span>    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">2.5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/quickpageHic-1.png" width="600" style="display: block; margin: auto;"></p>
<ol start="3" style="list-style-type: decimal">
<li>Annotate <code>plotgardener</code> plot objects by passing them into
the <code>plot</code> argument of annotation functions.</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/annoHeatmapLegend.html">annoHeatmapLegend</a></span><span class="op">(</span></span>
<span>    plot <span class="op">=</span> <span class="va">hicPlot</span>,</span>
<span>    x <span class="op">=</span> <span class="fl">2.85</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">0.1</span>, height <span class="op">=</span> <span class="fl">1.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/annoGenomeLabel.html">annoGenomeLabel</a></span><span class="op">(</span></span>
<span>    plot <span class="op">=</span> <span class="va">hicPlot</span>,</span>
<span>    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">2.75</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">0.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/quickpageAnno-1.png" width="600" style="display: block; margin: auto;"></p>
<p>For more information about how to place plots and annotations on a
<code>plotgardener</code> page, check out the section <a href="introduction_to_plotgardener.html#working-with-plot-objects-1">Working
with plot objects</a>.</p>
</div>
<div class="section level2">
<h2 id="exporting-plots">Exporting plots<a class="anchor" aria-label="anchor" href="#exporting-plots"></a>
</h2>
<p>When a <code>plotgardener</code> plot is ready to be saved and
exported, we can first remove all page guides that were made with
<code><a href="../reference/pageCreate.html">pageCreate()</a></code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pageGuideHide.html">pageGuideHide</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_to_plotgardener_files/figure-html/quickpageHide-1.png" width="600" style="display: block; margin: auto;"></p>
<p>We can then either use the <strong>Export</strong> toggle in the
RStudio plot panel, or save the plot within our R code as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/pageCreate.html">pageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span></span>
<span><span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHicSquare.html">plotHicSquare</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,</span>
<span>    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,</span>
<span>    assembly <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">2.5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/annoHeatmapLegend.html">annoHeatmapLegend</a></span><span class="op">(</span></span>
<span>    plot <span class="op">=</span> <span class="va">hicPlot</span>,</span>
<span>    x <span class="op">=</span> <span class="fl">2.85</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">0.1</span>, height <span class="op">=</span> <span class="fl">1.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/annoGenomeLabel.html">annoGenomeLabel</a></span><span class="op">(</span></span>
<span>    plot <span class="op">=</span> <span class="va">hicPlot</span>,</span>
<span>    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">2.75</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">0.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pageGuideHide.html">pageGuideHide</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><strong>Please note that due to the implementation of
<code>grid</code> removal functions,</strong> <strong>using
<code>pageGuideHide</code> within a <code>pdf</code> call will result in
the rendering of a</strong> <strong>separate, new page with the plot
guides removed. To avoid this artifact,</strong> <strong>hide guides in
the <code>pageCreate</code> function call with
<code>showGuides = FALSE</code>.</strong></p>
<p>For more detailed usage and examples, please refer to the other
available vignettes.</p>
</div>
<div class="section level2">
<h2 id="future-directions">Future Directions<a class="anchor" aria-label="anchor" href="#future-directions"></a>
</h2>
<p>We still have many ideas to add for a second version of
<code>plotgardener</code> including, but not limited to: grammar of
graphics style plot arguments and plot building, templates, themes, and
multi-plotting functions. If you have suggestions for ways we can
improve <code>plotgardener</code>, please let us know!</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    grid      stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] org.Hs.eg.db_3.15.0                    </span></span>
<span><span class="co">##  [2] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2</span></span>
<span><span class="co">##  [3] GenomicFeatures_1.48.3                 </span></span>
<span><span class="co">##  [4] AnnotationDbi_1.58.0                   </span></span>
<span><span class="co">##  [5] Biobase_2.56.0                         </span></span>
<span><span class="co">##  [6] GenomicRanges_1.48.0                   </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.32.2                    </span></span>
<span><span class="co">##  [8] IRanges_2.30.0                         </span></span>
<span><span class="co">##  [9] S4Vectors_0.34.0                       </span></span>
<span><span class="co">## [10] BiocGenerics_0.42.0                    </span></span>
<span><span class="co">## [11] plotgardenerData_1.2.0                 </span></span>
<span><span class="co">## [12] plotgardener_1.2.8                     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] bitops_1.0-7                matrixStats_0.62.0         </span></span>
<span><span class="co">##  [3] fs_1.5.2                    bit64_4.0.5                </span></span>
<span><span class="co">##  [5] filelock_1.0.2              progress_1.2.2             </span></span>
<span><span class="co">##  [7] httr_1.4.3                  RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [9] rprojroot_2.0.3             tools_4.2.0                </span></span>
<span><span class="co">## [11] bslib_0.4.0                 utf8_1.2.2                 </span></span>
<span><span class="co">## [13] R6_2.5.1                    DBI_1.1.3                  </span></span>
<span><span class="co">## [15] colorspace_2.0-3            prettyunits_1.1.1          </span></span>
<span><span class="co">## [17] tidyselect_1.1.2            bit_4.0.4                  </span></span>
<span><span class="co">## [19] curl_4.3.2                  compiler_4.2.0             </span></span>
<span><span class="co">## [21] textshaping_0.3.6           cli_3.3.0                  </span></span>
<span><span class="co">## [23] xml2_1.3.3                  desc_1.4.1                 </span></span>
<span><span class="co">## [25] DelayedArray_0.22.0         rtracklayer_1.56.1         </span></span>
<span><span class="co">## [27] sass_0.4.2                  scales_1.2.0               </span></span>
<span><span class="co">## [29] rappdirs_0.3.3              pkgdown_2.0.6              </span></span>
<span><span class="co">## [31] systemfonts_1.0.4           stringr_1.4.0              </span></span>
<span><span class="co">## [33] digest_0.6.29               Rsamtools_2.12.0           </span></span>
<span><span class="co">## [35] yulab.utils_0.0.5           rmarkdown_2.14             </span></span>
<span><span class="co">## [37] XVector_0.36.0              pkgconfig_2.0.3            </span></span>
<span><span class="co">## [39] htmltools_0.5.3             MatrixGenerics_1.8.1       </span></span>
<span><span class="co">## [41] dbplyr_2.2.1                fastmap_1.1.0              </span></span>
<span><span class="co">## [43] highr_0.9                   rlang_1.0.4                </span></span>
<span><span class="co">## [45] RSQLite_2.2.15              gridGraphics_0.5-1         </span></span>
<span><span class="co">## [47] jquerylib_0.1.4             BiocIO_1.6.0               </span></span>
<span><span class="co">## [49] generics_0.1.3              jsonlite_1.8.0             </span></span>
<span><span class="co">## [51] BiocParallel_1.30.3         dplyr_1.0.9                </span></span>
<span><span class="co">## [53] RCurl_1.98-1.7              magrittr_2.0.3             </span></span>
<span><span class="co">## [55] ggplotify_0.1.0             GenomeInfoDbData_1.2.8     </span></span>
<span><span class="co">## [57] Matrix_1.4-1                Rcpp_1.0.9                 </span></span>
<span><span class="co">## [59] munsell_0.5.0               fansi_1.0.3                </span></span>
<span><span class="co">## [61] lifecycle_1.0.1             stringi_1.7.8              </span></span>
<span><span class="co">## [63] yaml_2.3.5                  SummarizedExperiment_1.26.1</span></span>
<span><span class="co">## [65] zlibbioc_1.42.0             BiocFileCache_2.4.0        </span></span>
<span><span class="co">## [67] blob_1.2.3                  parallel_4.2.0             </span></span>
<span><span class="co">## [69] crayon_1.5.1                lattice_0.20-45            </span></span>
<span><span class="co">## [71] Biostrings_2.64.0           hms_1.1.1                  </span></span>
<span><span class="co">## [73] KEGGREST_1.36.3             knitr_1.39                 </span></span>
<span><span class="co">## [75] pillar_1.8.0                rjson_0.2.21               </span></span>
<span><span class="co">## [77] biomaRt_2.52.0              codetools_0.2-18           </span></span>
<span><span class="co">## [79] strawr_0.0.9                XML_3.99-0.10              </span></span>
<span><span class="co">## [81] glue_1.6.2                  evaluate_0.15              </span></span>
<span><span class="co">## [83] data.table_1.14.2           png_0.1-7                  </span></span>
<span><span class="co">## [85] vctrs_0.4.1                 gtable_0.3.0               </span></span>
<span><span class="co">## [87] purrr_0.3.4                 assertthat_0.2.1           </span></span>
<span><span class="co">## [89] cachem_1.0.6                ggplot2_3.3.6              </span></span>
<span><span class="co">## [91] xfun_0.31                   restfulr_0.0.15            </span></span>
<span><span class="co">## [93] ragg_1.2.2                  tibble_3.1.8               </span></span>
<span><span class="co">## [95] GenomicAlignments_1.32.1    plyranges_1.16.0           </span></span>
<span><span class="co">## [97] memoise_2.0.1               ellipsis_0.3.2</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nekramer" class="external-link">Nicole Kramer</a>, <a href="https://github.com/EricSDavis" class="external-link">Eric S. Davis</a>, <a href="https://github.com/cwenger" class="external-link">Craig Wenger</a>, <a href="https://github.com/dphansti" class="external-link">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
