<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to BentoBox • BentoBox</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to BentoBox">
<meta property="og:description" content="BentoBox">
<meta property="og:image" content="https://phanstiellab.github.io/bentobox/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BentoBox</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/our_philosophy.html">Our Philosophy</a>
</li>
<li>
  <a href="../articles/introduction_to_bentobox.html">Introduction to BentoBox</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/guides/bentobox_page.html">The BentoBox Page</a>
    </li>
    <li>
      <a href="../articles/guides/reading_data_for_bentobox.html">Reading Data for BentoBox</a>
    </li>
    <li>
      <a href="../articles/guides/plotting_multiomic_data.html">Plotting Multi-omic Data</a>
    </li>
    <li>
      <a href="../articles/guides/annotations.html">Plot Annotations</a>
    </li>
    <li>
      <a href="../articles/guides/bentobox_meta_functions.html">BentoBox Meta Functions</a>
    </li>
    <li>
      <a href="../articles/guides/bioconductor_integration.html">Bioconductor Integration</a>
    </li>
    <li>
      <a href="../articles/guides/incorporating_ggplots.html">Incorporating ggplots</a>
    </li>
    <li>
      <a href="../articles/guides/ideograms_and_shapes.html">Ideograms and Shapes</a>
    </li>
    <li>
      <a href="../articles/guides/aesthetics.html">Plot Aesthetics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/PhanstielLab/BentoBox">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="introduction_to_bentobox_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to BentoBox</h1>
                        <h4 class="author">Nicole Kramer, Eric S. Davis, Craig Wenger, Sarah Parker, Erika Deoudes, Douglas H. Phanstiel</h4>
            
            <h4 class="date">2021-07-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/PhanstielLab/BentoBox/blob/master/vignettes/introduction_to_bentobox.Rmd"><code>vignettes/introduction_to_bentobox.Rmd</code></a></small>
      <div class="hidden name"><code>introduction_to_bentobox.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p><img src="../man/figures/logo.png" align="right" width="140px" style="padding-left:20px; background-color:white; border-color: transparent"></p>
<p><code>BentoBox</code> is a coordinate-based, genomic visualization package for R. Using <code>grid</code> graphics, <code>BentoBox</code> empowers users to programatically and flexibly generate multi-panel figures. Tailored for genomics for a variety of genomic assemblies, <code>BentoBox</code> allows users to visualize large, complex genomic datasets while providing exquisite control over the arrangement of plots.</p>
<p><code>BentoBox</code> functions can be grouped into the following categories:</p>
<ul>
<li><strong>Page layout functions:</strong></li>
</ul>
<p>Functions for creating <code>BentoBox</code> page layouts, drawing, showing, and hiding guides, as well as placing plots on the page. See <a href="bentobox_page.html">The BentoBox Page</a></p>
<ul>
<li><strong>Reading functions:</strong></li>
</ul>
<p>Functions for quickly reading in large biological datasets. See <a href="reading_data_for_bentobox.html">Reading Data for BentoBox</a></p>
<ul>
<li><strong>Plotting functions:</strong></li>
</ul>
<p>Contains genomic plotting functions, functions for placing <code>ggplots</code> and <code>base</code> plots, as well as functions for drawing simple shapes. See <a href="plotting_multiomic_data.html">Plotting Multi-omic Data</a></p>
<ul>
<li><strong>Annotation functions:</strong></li>
</ul>
<p>Enables users to add annotations to their plots, such as legends, axes, and scales. See <a href="annotations.html">Plot Annotations</a></p>
<ul>
<li><strong>Meta functions:</strong></li>
</ul>
<p>Functions that display <code>BentoBox</code> properties or operate on other <code>BentoBox</code> functions, or constructors for <code>BentoBox</code> objects. See <a href="bentobox_meta_functions.html">BentoBox Meta Functions</a></p>
<p>This vignette provides a quick start guide for utilizing <code>BentoBox</code>. For in-depth demonstrations of <code>BentoBox</code>’s key features, see the additional articles. For detailed usage of each function, see the function-specific reference examples with <code>?function()</code> (e.g. <code>?bbPlotPairs()</code>).</p>
<p>All the data included in this vignette can be found in the supplementary package <code>BentoBoxData</code>.</p>
</div>
<div id="quick-plotting" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-plotting" class="anchor"></a>Quick plotting</h1>
<p><code>BentoBox</code> plotting functions contain 4 types of arguments:</p>
<ol style="list-style-type: decimal">
<li><p>Data reading argument (<code>data</code>)</p></li>
<li><p>Genomic locus arguments (<code>chrom</code>, <code>chromstart</code>, <code>chromend</code>, <code>assembly</code>)</p></li>
<li><p>Placement arguments (<code>x</code>, <code>y</code>, <code>width</code>, <code>height</code>, <code>just</code>, <code>default.units</code>, …) that define where each plot resides on a <code>bbPage</code></p></li>
<li><p>Attribute arguments that affect the data being plotted or the style of the plot (<code>norm</code>, <code>fill</code>, <code>fontcolor</code>, …) that vary between functions</p></li>
</ol>
<p>The quickest way to plot data is to omit the placement arguments. This will generate a <code>BentoBox</code> plot that fills up the entire graphics window and cannot be annotated. <strong>These plots are only meant to be used for quick</strong> <strong>genomic data inspection and not as final <code>BentoBox</code> plots.</strong> The only arguments that are required are the data arguments and locus arguments. The examples below show how to quickly plot different types of genomic data with plot defaults and included data types. To use your own data, replace the <code>data</code> argument with either a path to the file or an R object as described above.</p>
<div id="hi-c-matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#hi-c-matrices" class="anchor"></a>Hi-C matrices</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load BentoBox</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/BentoBox">BentoBox</a></span><span class="op">)</span>

<span class="co">## Load example Hi-C data</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/BentoBoxData">BentoBoxData</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span>

<span class="co">## Quick plot Hi-C data</span>
<span class="fu"><a href="../reference/bbPlotHicSquare.html">bbPlotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>
<span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/hic_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
<div id="signal-tracks" class="section level2">
<h2 class="hasAnchor">
<a href="#signal-tracks" class="anchor"></a>Signal tracks</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load BentoBox</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/BentoBox">BentoBox</a></span><span class="op">)</span>

<span class="co">## Load example signal data</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/BentoBoxData">BentoBoxData</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_ChIP_H3K27ac_signal"</span><span class="op">)</span>

<span class="co">## Quick plot signal data</span>
<span class="fu"><a href="../reference/bbPlotSignal.html">bbPlotSignal</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_ChIP_H3K27ac_signal</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>
<span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/signal_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
<div id="gene-tracks" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-tracks" class="anchor"></a>Gene tracks</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load BentoBox</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/BentoBox">BentoBox</a></span><span class="op">)</span>

<span class="co">## Load hg19 genomic annotation packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span>

<span class="co">## Quick plot genes</span>
<span class="fu"><a href="../reference/bbPlotGenes.html">bbPlotGenes</a></span><span class="op">(</span>
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>
<span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/gene_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
<div id="gwas-manhattan-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#gwas-manhattan-plots" class="anchor"></a>GWAS Manhattan plots</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load BentoBox</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://phanstiellab.github.io/BentoBox">BentoBox</a></span><span class="op">)</span>

<span class="co">## Load hg19 genomic annotation packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>

<span class="co">## Load example GWAS data</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PhanstielLab/BentoBoxData">BentoBoxData</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"hg19_insulin_GWAS"</span><span class="op">)</span>

<span class="co">## Quick plot GWAS data</span>
<span class="fu"><a href="../reference/bbPlotManhattan.html">bbPlotManhattan</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">hg19_insulin_GWAS</span>, 
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    fill <span class="op">=</span> <span class="fu"><a href="../reference/c.html">c</a></span><span class="op">(</span><span class="st">"steel blue"</span>, <span class="st">"grey"</span><span class="op">)</span>,
    ymax <span class="op">=</span> <span class="fl">1.1</span>, cex <span class="op">=</span> <span class="fl">0.20</span>
<span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/gwas_quickplot-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="plotting-and-annotating-on-the-bentobox-page" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting-and-annotating-on-the-bentobox-page" class="anchor"></a>Plotting and annotating on the <code>BentoBox</code> page</h1>
<p>To build complex, multi-panel <code>BentoBox</code> figures with annotations, we must:</p>
<ol style="list-style-type: decimal">
<li>Create a <code>BentoBox</code> coordinate page with <code><a href="../reference/bbPageCreate.html">bbPageCreate()</a></code>.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bbPageCreate.html">bbPageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/quickpage-1.png" width="600" style="display: block; margin: auto;"></p>
<ol start="2" style="list-style-type: decimal">
<li>Provide values for the placement arguments (<code>x</code>, <code>y</code>, <code>width</code>, <code>height</code>, <code>just</code>, <code>default.units</code>) in plotting functions and save the output of the plotting function.</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span>
<span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bbPlotHicSquare.html">bbPlotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">2.5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/quickpageHic-1.png" width="600" style="display: block; margin: auto;"></p>
<ol start="3" style="list-style-type: decimal">
<li>Annotate <code>BentoBox</code> plot objects by passing them into the <code>plot</code> argument of annotation functions.</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bbAnnoHeatmapLegend.html">bbAnnoHeatmapLegend</a></span><span class="op">(</span>
    plot <span class="op">=</span> <span class="va">hicPlot</span>,
    x <span class="op">=</span> <span class="fl">2.85</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">0.1</span>, height <span class="op">=</span> <span class="fl">1.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/bbAnnoGenomeLabel.html">bbAnnoGenomeLabel</a></span><span class="op">(</span>
    plot <span class="op">=</span> <span class="va">hicPlot</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">2.75</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">0.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/quickpageAnno-1.png" width="600" style="display: block; margin: auto;"></p>
<p>For more information about how to place plots and annotations on a <code>BentoBox</code> page, check out the section <a href="introduction_to_bentobox.html#working-with-plot-objects-1">Working with plot objects</a>.</p>
</div>
<div id="exporting-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#exporting-plots" class="anchor"></a>Exporting plots</h1>
<p>When a <code>BentoBox</code> plot is ready to be saved and exported, we can first remove all page guides that were made with <code><a href="../reference/bbPageCreate.html">bbPageCreate()</a></code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bbPageGuideHide.html">bbPageGuideHide</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="introduction_to_bentobox_files/figure-html/quickpageHide-1.png" width="600" style="display: block; margin: auto;"></p>
<p>We can then either use the <strong>Export</strong> toggle in the RStudio plot panel, or save the plot within our R code as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html">pdf</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span><span class="op">)</span>

<span class="fu"><a href="../reference/bbPageCreate.html">bbPageCreate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">3.25</span>, height <span class="op">=</span> <span class="fl">3.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"IMR90_HiC_10kb"</span><span class="op">)</span>
<span class="va">hicPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bbPlotHicSquare.html">bbPlotHicSquare</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="va">IMR90_HiC_10kb</span>,
    chrom <span class="op">=</span> <span class="st">"chr21"</span>, chromstart <span class="op">=</span> <span class="fl">28000000</span>, chromend <span class="op">=</span> <span class="fl">30300000</span>,
    assembly <span class="op">=</span> <span class="st">"hg19"</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">2.5</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>
<span class="fu"><a href="../reference/bbAnnoHeatmapLegend.html">bbAnnoHeatmapLegend</a></span><span class="op">(</span>
    plot <span class="op">=</span> <span class="va">hicPlot</span>,
    x <span class="op">=</span> <span class="fl">2.85</span>, y <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">0.1</span>, height <span class="op">=</span> <span class="fl">1.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>

<span class="fu"><a href="../reference/bbAnnoGenomeLabel.html">bbAnnoGenomeLabel</a></span><span class="op">(</span>
    plot <span class="op">=</span> <span class="va">hicPlot</span>,
    x <span class="op">=</span> <span class="fl">0.25</span>, y <span class="op">=</span> <span class="fl">2.75</span>, width <span class="op">=</span> <span class="fl">2.5</span>, height <span class="op">=</span> <span class="fl">0.25</span>, default.units <span class="op">=</span> <span class="st">"inches"</span>
<span class="op">)</span>
<span class="fu"><a href="../reference/bbPageGuideHide.html">bbPageGuideHide</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>For more detailed usage and examples, please refer to the other available vignettes.</p>
</div>
<div id="future-directions" class="section level1">
<h1 class="hasAnchor">
<a href="#future-directions" class="anchor"></a>Future Directions</h1>
<p>We still have many ideas to add for a second version of <code>BentoBox</code> including, but not limited to: grammar of graphics style plot arguments and plot building, templates, themes, and multi-plotting functions. If you have suggestions for ways we can improve <code>BentoBox</code>, please let us know!</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt;  [1] stats4    parallel  grid      stats     graphics  grDevices utils    </span>
<span class="co">#&gt;  [8] datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] org.Hs.eg.db_3.13.0                    </span>
<span class="co">#&gt;  [2] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2</span>
<span class="co">#&gt;  [3] GenomicFeatures_1.44.0                 </span>
<span class="co">#&gt;  [4] AnnotationDbi_1.54.1                   </span>
<span class="co">#&gt;  [5] Biobase_2.52.0                         </span>
<span class="co">#&gt;  [6] GenomicRanges_1.44.0                   </span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.28.1                    </span>
<span class="co">#&gt;  [8] IRanges_2.26.0                         </span>
<span class="co">#&gt;  [9] S4Vectors_0.30.0                       </span>
<span class="co">#&gt; [10] BiocGenerics_0.38.0                    </span>
<span class="co">#&gt; [11] BentoBoxData_0.99.0                    </span>
<span class="co">#&gt; [12] BentoBox_0.99.6                        </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-7                matrixStats_0.60.0         </span>
<span class="co">#&gt;  [3] fs_1.5.0                    bit64_4.0.5                </span>
<span class="co">#&gt;  [5] filelock_1.0.2              progress_1.2.2             </span>
<span class="co">#&gt;  [7] httr_1.4.2                  RColorBrewer_1.1-2         </span>
<span class="co">#&gt;  [9] rprojroot_2.0.2             tools_4.1.0                </span>
<span class="co">#&gt; [11] bslib_0.2.5.1               utf8_1.2.2                 </span>
<span class="co">#&gt; [13] R6_2.5.0                    DBI_1.1.1                  </span>
<span class="co">#&gt; [15] colorspace_2.0-2            prettyunits_1.1.1          </span>
<span class="co">#&gt; [17] tidyselect_1.1.1            bit_4.0.4                  </span>
<span class="co">#&gt; [19] curl_4.3.2                  compiler_4.1.0             </span>
<span class="co">#&gt; [21] textshaping_0.3.5           xml2_1.3.2                 </span>
<span class="co">#&gt; [23] desc_1.3.0                  DelayedArray_0.18.0        </span>
<span class="co">#&gt; [25] rtracklayer_1.52.0          sass_0.4.0                 </span>
<span class="co">#&gt; [27] scales_1.1.1                rappdirs_0.3.3             </span>
<span class="co">#&gt; [29] pkgdown_1.6.1               systemfonts_1.0.2          </span>
<span class="co">#&gt; [31] stringr_1.4.0               digest_0.6.27              </span>
<span class="co">#&gt; [33] Rsamtools_2.8.0             rmarkdown_2.9              </span>
<span class="co">#&gt; [35] XVector_0.32.0              pkgconfig_2.0.3            </span>
<span class="co">#&gt; [37] htmltools_0.5.1.1           MatrixGenerics_1.4.0       </span>
<span class="co">#&gt; [39] dbplyr_2.1.1                fastmap_1.1.0              </span>
<span class="co">#&gt; [41] highr_0.9                   rlang_0.4.11               </span>
<span class="co">#&gt; [43] RSQLite_2.2.7               gridGraphics_0.5-1         </span>
<span class="co">#&gt; [45] jquerylib_0.1.4             BiocIO_1.2.0               </span>
<span class="co">#&gt; [47] generics_0.1.0              jsonlite_1.7.2             </span>
<span class="co">#&gt; [49] BiocParallel_1.26.1         dplyr_1.0.7                </span>
<span class="co">#&gt; [51] RCurl_1.98-1.3              magrittr_2.0.1             </span>
<span class="co">#&gt; [53] ggplotify_0.0.7             GenomeInfoDbData_1.2.6     </span>
<span class="co">#&gt; [55] Matrix_1.3-4                Rcpp_1.0.7                 </span>
<span class="co">#&gt; [57] munsell_0.5.0               fansi_0.5.0                </span>
<span class="co">#&gt; [59] lifecycle_1.0.0             stringi_1.7.3              </span>
<span class="co">#&gt; [61] yaml_2.2.1                  SummarizedExperiment_1.22.0</span>
<span class="co">#&gt; [63] zlibbioc_1.38.0             BiocFileCache_2.0.0        </span>
<span class="co">#&gt; [65] blob_1.2.2                  crayon_1.4.1               </span>
<span class="co">#&gt; [67] lattice_0.20-44             Biostrings_2.60.1          </span>
<span class="co">#&gt; [69] hms_1.1.0                   KEGGREST_1.32.0            </span>
<span class="co">#&gt; [71] knitr_1.33                  pillar_1.6.2               </span>
<span class="co">#&gt; [73] rjson_0.2.20                biomaRt_2.48.2             </span>
<span class="co">#&gt; [75] strawr_0.0.8                XML_3.99-0.6               </span>
<span class="co">#&gt; [77] glue_1.4.2                  evaluate_0.14              </span>
<span class="co">#&gt; [79] data.table_1.14.0           BiocManager_1.30.16        </span>
<span class="co">#&gt; [81] png_0.1-7                   vctrs_0.3.8                </span>
<span class="co">#&gt; [83] gtable_0.3.0                purrr_0.3.4                </span>
<span class="co">#&gt; [85] assertthat_0.2.1            cachem_1.0.5               </span>
<span class="co">#&gt; [87] ggplot2_3.3.5               xfun_0.24                  </span>
<span class="co">#&gt; [89] restfulr_0.0.13             ragg_1.1.3                 </span>
<span class="co">#&gt; [91] tibble_3.1.3                rvcheck_0.1.8              </span>
<span class="co">#&gt; [93] GenomicAlignments_1.28.0    plyranges_1.12.1           </span>
<span class="co">#&gt; [95] memoise_2.0.0               ellipsis_0.3.2</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/nekramer">Nicole Kramer</a>, <a href="https://github.com/EricSDavis">Eric S. Davis</a>, <a href="https://github.com/cwenger">Craig Wenger</a>, <a href="https://github.com/dphansti">Douglas H. Phanstiel</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
